{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BreCaNet\n",
    "\n",
    "This Jupyter notebook goes through the steps to process the full data sets before running PANDA\n",
    "\n",
    "##### Gene Expression Data\n",
    "\n",
    "HTSeq\n",
    "FPKM\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import shutil\n",
    "import csv\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Paths to expression data\n",
    "\n",
    "dataPath_HTSeq = '/Users/ursulawidocki/Desktop/BreCaNet/Data/gdc_download_HTSeq' #zipped file path\n",
    "dataNewFolder_HTSeq = '/Users/ursulawidocki/Desktop/BreCaNet/Data/processedTCGA_HTSeq'\n",
    "\n",
    "dataPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/gdc_download_FPKM_832' #zipped file path\n",
    "dataNewFolder = '/Users/ursulawidocki/Desktop/BreCaNet/Data/processedTCGA_832' # folder for unzipped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Unzips HTSeq expression data\n",
    "# creates a folder (processedTCGA_HTSeq) of the unzipped files \n",
    "# Gets the codes of the patient folders and expression files\n",
    "\n",
    "folderIDs_HTSeq = list() # codes of the patient IDs on the folder\n",
    "indivList_HTSeq = list() # codes of the patients' expression file\n",
    "\n",
    "for fileName in os.listdir(dataPath_HTSeq):\n",
    "    if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "        \n",
    "    else:\n",
    "        folderIDs_HTSeq.append(fileName)\n",
    "        \n",
    "    newPath = dataPath_HTSeq + '/' + fileName\n",
    "    \n",
    "    # goes into the individual folder\n",
    "    for indivName in os.listdir(newPath):\n",
    "        if (indivName == '.DS_Store') or (indivName == 'MANIFEST.txt') or (indivName == 'annotations.txt'):\n",
    "            continue\n",
    "        filePath = newPath + '/' + indivName\n",
    "        \n",
    "        fileEnding = indivName[-3] + indivName[-2] + indivName[-1]\n",
    "        if(fileEnding == '.gz'):\n",
    "            with gzip.open(filePath, 'rb') as f_in:\n",
    "                newFilePath = dataNewFolder_HTSeq + '/' + indivName[0:-3]\n",
    "                indivList_HTSeq.append(indivName[0:-3])\n",
    "                \n",
    "                with open(newFilePath, 'wb') as f_out:\n",
    "                    shutil.copyfileobj(f_in, f_out)\n",
    "                    \n",
    "                    \n",
    "## Unzips FPKM expression data\n",
    "# creates a folder (processedTCGA_FPKM) of the unzipped files \n",
    "# Gets the codes of the patient folders and expression files\n",
    "\n",
    "folderIDs = list() # codes of the patient IDs on the folder\n",
    "indivList = list() # codes of the patients' expression file\n",
    "\n",
    "for fileName in os.listdir(dataPath):\n",
    "    if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "        \n",
    "    else:\n",
    "        folderIDs.append(fileName)\n",
    "        \n",
    "    newPath = dataPath + '/' + fileName\n",
    "    \n",
    "    # goes into the individual folder\n",
    "    for indivName in os.listdir(newPath):\n",
    "        if (indivName == '.DS_Store') or (indivName == 'MANIFEST.txt') or (indivName == 'annotations.txt'):\n",
    "            continue\n",
    "        filePath = newPath + '/' + indivName\n",
    "        \n",
    "        fileEnding = indivName[-3] + indivName[-2] + indivName[-1]\n",
    "        if(fileEnding == '.gz'):\n",
    "            with gzip.open(filePath, 'rb') as f_in:\n",
    "                newFilePath = dataNewFolder + '/' + indivName[0:-3]\n",
    "                indivList.append(indivName[0:-3])\n",
    "                \n",
    "                with open(newFilePath, 'wb') as f_out:\n",
    "                    shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "## Checks if folder IDs and file IDs are similar between the two \n",
    "# expression sets\n",
    "\n",
    "print(len(folderIDs_HTSeq) == len(folderIDs))\n",
    "print(len(indivList_HTSeq) == len(indivList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes file of folder names\n",
    "\n",
    "folderIDPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/foldIDList_HTSeq.txt'\n",
    "folderFile = open(folderIDPath, \"w+\")\n",
    "for i in range(0, 832):\n",
    "    newLineFolder = folderIDs_HTSeq[i] + \"\\n\"\n",
    "    \n",
    "    folderFile.write(newLineFolder)\n",
    "    \n",
    "folderFile.close()\n",
    "\n",
    "folderIDPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/foldIDList_FPKM_832.txt'\n",
    "folderFile = open(folderIDPath, \"w+\")\n",
    "for i in range(0, 832):\n",
    "    newLineFolder = folderIDs[i] + \"\\n\"\n",
    "    \n",
    "    folderFile.write(newLineFolder)\n",
    "    \n",
    "folderFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Gets folder codes from expression data\n",
    "\n",
    "folderIDPath_HTSeq = '/Users/ursulawidocki/Desktop/BreCaNet/Data/foldIDList_HTSeq.txt'\n",
    "folders_HTSeq = pd.read_csv(folderIDPath_HTSeq, delimiter = '\\t', header)\n",
    "folders_list_HTSeq = list(folders_HTSeq[0])\n",
    "\n",
    "folderIDPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/foldIDList_FPKM_832.txt'\n",
    "folders = pd.read_csv(folderIDPath, delimiter = '\\t', header = None)\n",
    "folders_list = list(folders[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "714\n",
      "714\n",
      "714\n",
      "714\n",
      "714\n",
      "714\n",
      "714\n",
      "714\n",
      "False\n",
      "False\n",
      "False\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "## Reads in metadata\n",
    "# reads in case_ids such that only the first instance is used\n",
    "\n",
    "# HTSeq\n",
    "# gets file_id_HTSeq and case_id_HTSeq\n",
    "metaPath_HTSeq = '/Users/ursulawidocki/Desktop/BreCaNet/Data/metadata.cart.HTSeq.json'\n",
    "with open(metaPath_HTSeq, 'r') as f:\n",
    "    meta_dict_HTSeq = json.load(f)\n",
    "\n",
    "file_id_HTSeq = list()\n",
    "case_id_HTSeq = list()\n",
    "for i in range(0, len(meta_dict_HTSeq)):\n",
    "    if(meta_dict_HTSeq[i].get(\"associated_entities\")[0].get(\"case_id\") not in case_id_HTSeq):\n",
    "        file_id_HTSeq.append(meta_dict_HTSeq[i].get(\"file_id\"))\n",
    "        case_id_HTSeq.append(meta_dict_HTSeq[i].get(\"associated_entities\")[0].get(\"case_id\"))\n",
    "            \n",
    "\n",
    "print(len(file_id_HTSeq))\n",
    "print(len(case_id_HTSeq))\n",
    "print(len(set(file_id_HTSeq)))\n",
    "print(len(set(case_id_HTSeq)))\n",
    "\n",
    "# FPKM\n",
    "# gets file_id and case_id\n",
    "metaPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/metadata.cart.FPKM_832.json'\n",
    "with open(metaPath, 'r') as f:\n",
    "    meta_dict = json.load(f)\n",
    "    \n",
    "file_id = list()\n",
    "case_id = list()\n",
    "for i in range(0, len(meta_dict)):\n",
    "    if(meta_dict[i].get(\"associated_entities\")[0].get(\"case_id\") not in case_id):\n",
    "        file_id.append(meta_dict[i].get(\"file_id\"))\n",
    "        case_id.append(meta_dict[i].get(\"associated_entities\")[0].get(\"case_id\"))\n",
    "\n",
    "print(len(file_id))\n",
    "print(len(case_id))\n",
    "print(len(set(file_id)))\n",
    "print(len(set(case_id)))\n",
    "\n",
    "print(file_id_HTSeq == file_id)\n",
    "print(set(file_id_HTSeq) == set(file_id))\n",
    "print(case_id_HTSeq == case_id)\n",
    "print(set(case_id_HTSeq) == set(case_id)) # since True, these two datasets are of same samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1097\n",
      "1097\n"
     ]
    }
   ],
   "source": [
    "## Gets uuid and barcodes of the patients from clinical data\n",
    "\n",
    "# reads in clinical information to link patients with their barcode\n",
    "clinicalPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/gdc_download_clinical_832/8162d394-8b64-4da2-9f5b-d164c54b9608/nationwidechildrens.org_clinical_patient_brca.txt'\n",
    "clinical_df = pd.read_csv(clinicalPath, delimiter = '\\t')\n",
    "\n",
    "# removes rows that are not necessary to match data\n",
    "clinical_df = clinical_df.drop(clinical_df.index[0])\n",
    "clinical_df = clinical_df.drop(clinical_df.index[0])\n",
    "clinical_df.head()\n",
    "codes_df = clinical_df[['bcr_patient_uuid', 'bcr_patient_barcode']]\n",
    "#codes_df.head()\n",
    "\n",
    "uuid_list = list(codes_df['bcr_patient_uuid'])\n",
    "barcode_list = list(codes_df['bcr_patient_barcode'])\n",
    "\n",
    "print(len(uuid_list))\n",
    "print(len(barcode_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1087\n",
      "1087\n"
     ]
    }
   ],
   "source": [
    "## Reads in PAM50 labels\n",
    "\n",
    "subtype_path = \"/Users/ursulawidocki/Desktop/BreCaNet/Code/TCGA_subtypes.txt\"\n",
    "subtypes = pd.read_csv(subtype_path, delimiter = \" \")\n",
    "subtypes.head()\n",
    "\n",
    "pt_barcodes_list = list(subtypes['brca_subtypes.patient'])\n",
    "subtype_list = list(subtypes['brca_subtypes.BRCA_Subtype_PAM50'])\n",
    "\n",
    "print(len(pt_barcodes_list))\n",
    "print(len(subtype_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "714\n",
      "714\n"
     ]
    }
   ],
   "source": [
    "## First, match the folder with the file_ids to get the corresponding cases from the metadata\n",
    "\n",
    "match_cases_HTSeq = list()\n",
    "\n",
    "for folder in folders_list_HTSeq:\n",
    "    if folder in file_id_HTSeq:\n",
    "        ind = file_id_HTSeq.index(folder)\n",
    "        match_cases_HTSeq.append(case_id_HTSeq[ind])\n",
    "\n",
    "\n",
    "## FPKM\n",
    "\n",
    "match_cases = list()\n",
    "\n",
    "for folder in folders_list:\n",
    "    if folder in file_id:\n",
    "        ind = file_id.index(folder)\n",
    "        match_cases.append(case_id[ind])\n",
    "\n",
    "print(set(match_cases_HTSeq) == set(match_cases))\n",
    "print(len(match_cases_HTSeq))\n",
    "print(len(match_cases))\n",
    "\n",
    "# there are 14 file_ids that cannot be matched to expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "714\n",
      "714\n"
     ]
    }
   ],
   "source": [
    "## Matches case_ids from metadata with uuid and patient barcodes from clinical data\n",
    "\n",
    "match_barcodes_HTSeq = list()\n",
    "\n",
    "for case in match_cases_HTSeq:\n",
    "    for ptID in uuid_list:\n",
    "        if case == ptID.lower():\n",
    "            ind = uuid_list.index(ptID)\n",
    "            match_barcodes_HTSeq.append(barcode_list[ind])\n",
    "            break\n",
    "\n",
    "\n",
    "match_barcodes = list()\n",
    "\n",
    "for case in match_cases:\n",
    "    for ptID in uuid_list:\n",
    "        if case == ptID.lower():\n",
    "            ind = uuid_list.index(ptID)\n",
    "            match_barcodes.append(barcode_list[ind])\n",
    "            break\n",
    "            \n",
    "print(set(match_barcodes_HTSeq) == set(match_barcodes))\n",
    "print(len(match_barcodes_HTSeq))\n",
    "print(len(match_barcodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "714\n",
      "714\n"
     ]
    }
   ],
   "source": [
    "## Matches subtypes to their barcodes from the TCGA data\n",
    "\n",
    "# HTSeq\n",
    "match_subtypes_HTSeq = list()\n",
    "\n",
    "for barcode in match_barcodes_HTSeq:\n",
    "    ind = pt_barcodes_list.index(barcode)\n",
    "    match_subtypes_HTSeq.append(subtype_list[ind])\n",
    "\n",
    "\n",
    "# FPKM\n",
    "match_subtypes = list()\n",
    "\n",
    "for barcode in match_barcodes:\n",
    "    ind = pt_barcodes_list.index(barcode)\n",
    "    match_subtypes.append(subtype_list[ind])\n",
    "\n",
    "print(set(match_subtypes_HTSeq) == set(match_subtypes))\n",
    "print(len(match_subtypes_HTSeq))\n",
    "print(len(match_subtypes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "{'Her2', 'Basal', 'LumB', 'Normal', 'LumA'}\n"
     ]
    }
   ],
   "source": [
    "print(len(set(match_subtypes))) # there is only this many subtypes among my samples\n",
    "print((set(match_subtypes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "60483\n",
      "60483\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "## gets list of genes that are in the file\n",
    "\n",
    "# reads in HTSeq\n",
    "geneList_HTSeq = list() # list of all genes that are noted in the expression files\n",
    "\n",
    "with open(dataNewFolder_HTSeq + '/' + indivList_HTSeq[0], \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        potentialGene = line.strip().split(\"\\t\")[0]\n",
    "        \n",
    "        if potentialGene[0:3] == \"ENS\":\n",
    "            newGeneName = potentialGene.strip().split(\".\")[0]\n",
    "            geneList_HTSeq.append(newGeneName)\n",
    "print(len(geneList_HTSeq))\n",
    "\n",
    "\n",
    "# reads in FPKM\n",
    "geneList = list() # list of all genes that are noted in the expression files\n",
    "\n",
    "with open(dataNewFolder + '/' + indivList[0], \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        potentialGene = line.strip().split(\"\\t\")[0]\n",
    "        \n",
    "        if potentialGene[0:3] == \"ENS\":\n",
    "            newGeneName = potentialGene.strip().split(\".\")[0]\n",
    "            geneList.append(newGeneName)\n",
    "print(len(geneList))\n",
    "\n",
    "geneList_lumA = geneList\n",
    "geneList_lumB = geneList\n",
    "geneList_basal = geneList\n",
    "geneList_her2 = geneList\n",
    "geneList_norm = geneList\n",
    "\n",
    "print(set(geneList_HTSeq) == set(geneList)) # note: genes not in same order in files but same genes present"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing raw counts (HTSeq) for DE analysis (done in R)\n",
    "\n",
    "\n",
    "#### Now, sort the folder IDs into their subtypes\n",
    "Notice that the order of folders_list_HTSeq is in the same order as match_subtypes_HTSeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## uses index of subtype label to sort folder IDs into subtypes\n",
    "\n",
    "# Luminal A samples\n",
    "lumA_samples_HTSeq = list() # folder name of expression files\n",
    "lumA_indiv_HTSeq = list() # labels of expression files\n",
    "for i in range(0,len(match_subtypes_HTSeq)):\n",
    "    if match_subtypes_HTSeq[i]==\"LumA\":\n",
    "        lumA_samples_HTSeq.append(file_id_HTSeq[i])\n",
    "        lumA_indiv_HTSeq.append(indivList_HTSeq[folderIDs_HTSeq.index(file_id_HTSeq[i])])\n",
    "        \n",
    "# Luminal B samples\n",
    "lumB_samples_HTSeq = list()\n",
    "lumB_indiv_HTSeq = list()\n",
    "for i in range(0,len(match_subtypes_HTSeq)):\n",
    "    if match_subtypes_HTSeq[i]==\"LumB\":\n",
    "        lumB_samples_HTSeq.append(file_id_HTSeq[i])\n",
    "        lumB_indiv_HTSeq.append(indivList_HTSeq[folderIDs_HTSeq.index(file_id_HTSeq[i])])\n",
    "        \n",
    "# Basal samples\n",
    "basal_samples_HTSeq = list()\n",
    "basal_indiv_HTSeq = list()\n",
    "for i in range(0,len(match_subtypes_HTSeq)):\n",
    "    if match_subtypes_HTSeq[i]==\"Basal\":\n",
    "        basal_samples_HTSeq.append(file_id_HTSeq[i])\n",
    "        basal_indiv_HTSeq.append(indivList_HTSeq[folderIDs_HTSeq.index(file_id_HTSeq[i])])\n",
    "        \n",
    "# Her2 samples\n",
    "her2_samples_HTSeq = list()\n",
    "her2_indiv_HTSeq = list()\n",
    "for i in range(0,len(match_subtypes_HTSeq)):\n",
    "    if match_subtypes[i]==\"Her2\":\n",
    "        her2_samples_HTSeq.append(file_id_HTSeq[i])\n",
    "        her2_indiv_HTSeq.append(indivList_HTSeq[folderIDs_HTSeq.index(file_id_HTSeq[i])])\n",
    "\n",
    "# Normal samples\n",
    "norm_samples_HTSeq = list()\n",
    "norm_indiv_HTSeq = list()\n",
    "for i in range(0,len(match_subtypes_HTSeq)):\n",
    "    if match_subtypes_HTSeq[i]==\"Normal\":\n",
    "        norm_samples_HTSeq.append(file_id_HTSeq[i])\n",
    "        norm_indiv_HTSeq.append(indivList_HTSeq[folderIDs_HTSeq.index(file_id_HTSeq[i])])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60483, 714)\n"
     ]
    }
   ],
   "source": [
    "## Makes a giant expression matrix for edgeR\n",
    "# makes a matrix where patients are columns and genes are rows\n",
    "\n",
    "indivList_HTSeq2 =lumA_indiv_HTSeq + lumB_indiv_HTSeq + basal_indiv_HTSeq + her2_indiv_HTSeq + norm_indiv_HTSeq\n",
    "all_exp_HTSeq = np.zeros((len(geneList_HTSeq), len(indivList_HTSeq2))) # matrix with all info from subtypes\n",
    "j = 0 # index of the column (sample)\n",
    "i = 0 # index of the row (gene)\n",
    "print(all_exp_HTSeq.shape) # just to make sure dimensions are correct\n",
    "\n",
    "for fileName in os.listdir(dataNewFolder_HTSeq):\n",
    "    if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "        \n",
    "    if fileName in indivList_HTSeq2:\n",
    "        temp_folder = file_id_HTSeq[indivList_HTSeq2.index(fileName)]\n",
    "        readFile = dataNewFolder_HTSeq + '/' + fileName\n",
    "    \n",
    "        with open(readFile, \"r\") as file:\n",
    "            for line in file.readlines():\n",
    "                if(\"no_feature\" in line):\n",
    "                        break\n",
    "                all_exp_HTSeq[i, j] = float(line.strip().split(\"\\t\")[1])\n",
    "                i+=1\n",
    "                \n",
    "            i = 0\n",
    "            if j != len(indivList_HTSeq2):\n",
    "                j+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[2.527e+03 1.397e+03 4.948e+03 ... 0.000e+00 0.000e+00 0.000e+00]\n",
      " [3.000e+00 2.000e+00 2.100e+01 ... 0.000e+00 0.000e+00 0.000e+00]\n",
      " [2.413e+03 2.021e+03 1.668e+03 ... 0.000e+00 0.000e+00 0.000e+00]\n",
      " [3.410e+03 2.140e+03 2.220e+03 ... 0.000e+00 0.000e+00 0.000e+00]\n",
      " [1.281e+03 7.480e+02 6.120e+02 ... 0.000e+00 0.000e+00 0.000e+00]]\n"
     ]
    }
   ],
   "source": [
    "print(all_exp_HTSeq[0:5,])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes a reference file of file names and their subtypes\n",
    "\n",
    "ref_col_path = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/all_labels.txt'\n",
    "refFile = open(ref_col_path,\"w+\")\n",
    "\n",
    "for i in range(0, len(match_barcodes_HTSeq)):\n",
    "    ref_row = match_barcodes_HTSeq[i] + \"\\t\" + match_subtypes_HTSeq[i] + \"\\n\"\n",
    "    refFile.write(ref_row)\n",
    "    \n",
    "refFile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing FPKM data for PANDA networks\n",
    "\n",
    "\n",
    "#### Now, sort the folder IDs into their subtypes\n",
    "Notice that the order of folders_list is in the same order as match_subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "## uses index of subtype label to sort folder IDs into subtypes\n",
    "\n",
    "# Luminal A samples\n",
    "lumA_samples = list()\n",
    "lumA_indiv = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"LumA\":\n",
    "        lumA_samples.append(folders_list[i])\n",
    "        lumA_indiv.append(indivList[folderIDs.index(file_id[i])])\n",
    "\n",
    "# Luminal B samples\n",
    "lumB_samples = list()\n",
    "lumB_indiv = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"LumB\":\n",
    "        lumB_samples.append(folders_list[i])\n",
    "        lumB_indiv.append(indivList[folderIDs.index(file_id[i])])\n",
    "\n",
    "# Basal samples\n",
    "basal_samples = list()\n",
    "basal_indiv = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"Basal\":\n",
    "        basal_samples.append(folders_list[i])\n",
    "        basal_indiv.append(indivList[folderIDs.index(file_id[i])])\n",
    "\n",
    "# Her2 samples\n",
    "her2_samples = list()\n",
    "her2_indiv = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"Her2\":\n",
    "        her2_samples.append(folders_list[i])\n",
    "        her2_indiv.append(indivList[folderIDs.index(file_id[i])])\n",
    "\n",
    "# Normal samples\n",
    "norm_samples = list()\n",
    "norm_indiv = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"Normal\":\n",
    "        norm_samples.append(folders_list[i])\n",
    "        norm_indiv.append(indivList[folderIDs.index(file_id[i])])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60483, 26)\n"
     ]
    }
   ],
   "source": [
    "## Reads through the unzipped expression files and creates the expression matrices depending on how the file\n",
    "# name matches with the folder IDs of each subtype\n",
    "\n",
    "lumA_exp = np.zeros((len(geneList), len(lumA_samples))) # exp matrix of LumA samples\n",
    "lumB_exp = np.zeros((len(geneList), len(lumB_samples))) # exp matrix of LumB samples\n",
    "basal_exp = np.zeros((len(geneList), len(basal_samples))) # exp matrix of basal samples\n",
    "her2_exp = np.zeros((len(geneList), len(her2_samples))) # exp matrix of her2 samples\n",
    "norm_exp = np.zeros((len(geneList), len(norm_samples))) # exp matrix of LumA samples\n",
    "indivList2 = lumA_indiv + lumB_indiv + basal_indiv + her2_indiv + norm_indiv\n",
    "\n",
    "print(norm_exp.shape) # just to make sure dimensions are correct\n",
    " \n",
    "lumA_j = 0 # index of the column (sample)index of the column (sample)\n",
    "lumB_j = 0\n",
    "basal_j = 0\n",
    "her2_j = 0\n",
    "norm_j = 0\n",
    "for fileName in os.listdir(dataNewFolder):\n",
    "    if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "    \n",
    "    if fileName in indivList2:\n",
    "        \n",
    "        temp_folder = folderIDs[indivList2.index(fileName)]\n",
    "        readFile = dataNewFolder + '/' + fileName\n",
    "            \n",
    "        # adds to luminal A expression matrix\n",
    "        if temp_folder in lumA_samples:\n",
    "            \n",
    "            lumA_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    lumA_exp[lumA_i, lumA_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    lumA_i += 1\n",
    "                    \n",
    "                lumA_j += 1\n",
    "                        \n",
    "        # adds to luminal B expression matrix\n",
    "        if temp_folder in lumB_samples:\n",
    "            \n",
    "            lumB_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    lumB_exp[lumB_i, lumB_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    lumB_i += 1\n",
    "                    \n",
    "                lumB_j += 1\n",
    "        \n",
    "        # adds to basal expression matrix\n",
    "        if temp_folder in basal_samples:\n",
    "            \n",
    "            basal_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    basal_exp[basal_i, basal_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    basal_i += 1\n",
    "                    \n",
    "                basal_j += 1\n",
    "        \n",
    "        # adds to her2 expression matrix\n",
    "        if temp_folder in her2_samples:\n",
    "            \n",
    "            her2_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    her2_exp[her2_i, her2_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    her2_i += 1\n",
    "                    \n",
    "                her2_j += 1\n",
    "        \n",
    "        # adds to norm expression matrix\n",
    "        if temp_folder in norm_samples:\n",
    "            \n",
    "            norm_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    norm_exp[norm_i, norm_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    norm_i += 1\n",
    "                    \n",
    "                norm_j += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(50137, 26)\n",
      "OLD  60483\n",
      "ZEROS  10346\n",
      "NEW  50137\n",
      "50137\n"
     ]
    }
   ],
   "source": [
    "## Removes zero rows from expression data\n",
    "\n",
    "# Luminal A data\n",
    "rowList_lumA = list()\n",
    "for i in range(0, len(lumA_exp)):\n",
    "    if np.sum(lumA_exp[i,:]) == 0.0:\n",
    "        rowList_lumA.append(i)\n",
    "\n",
    "lumA_exp2 = np.zeros(((len(lumA_exp)-len(rowList_lumA)), len(lumA_samples))) # matrix without zero rows\n",
    "geneList_lumA2 = list()\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(lumA_exp)):\n",
    "    if j in rowList_lumA:\n",
    "        j+=1\n",
    "    else:\n",
    "        lumA_exp2[r,:] = lumA_exp[j,:]\n",
    "        geneList_lumA2.append(geneList_lumA[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "# Luminal B data\n",
    "rowList_lumB = list()\n",
    "for i in range(0, len(lumB_exp)):\n",
    "    if np.sum(lumB_exp[i,:]) == 0.0:\n",
    "        rowList_lumB.append(i)\n",
    "\n",
    "lumB_exp2 = np.zeros(((len(lumB_exp)-len(rowList_lumB)), len(lumB_samples))) # matrix without zero rows\n",
    "geneList_lumB2 = list()\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(lumB_exp)):\n",
    "    if j in rowList_lumB:\n",
    "        j+=1\n",
    "    else:\n",
    "        lumB_exp2[r,:] = lumB_exp[j,:]\n",
    "        geneList_lumB2.append(geneList_lumB[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "# Basal data\n",
    "rowList_basal = list()\n",
    "for i in range(0, len(basal_exp)):\n",
    "    if np.sum(basal_exp[i,:]) == 0.0:\n",
    "        rowList_basal.append(i)\n",
    "\n",
    "basal_exp2 = np.zeros(((len(basal_exp)-len(rowList_basal)), len(basal_samples))) # matrix without zero rows\n",
    "geneList_basal2 = list()\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(basal_exp)):\n",
    "    if j in rowList_basal:\n",
    "        j+=1\n",
    "    else:\n",
    "        basal_exp2[r,:] = basal_exp[j,:]\n",
    "        geneList_basal2.append(geneList_basal[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "# HER2 data\n",
    "rowList_her2 = list()\n",
    "for i in range(0, len(her2_exp)):\n",
    "    if np.sum(her2_exp[i,:]) == 0.0:\n",
    "        rowList_her2.append(i)\n",
    "\n",
    "her2_exp2 = np.zeros(((len(her2_exp)-len(rowList_her2)), len(her2_samples))) # matrix without zero rows\n",
    "geneList_her2_2 = list()\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(her2_exp)):\n",
    "    if j in rowList_her2:\n",
    "        j+=1\n",
    "    else:\n",
    "        her2_exp2[r,:] = her2_exp[j,:]\n",
    "        geneList_her2_2.append(geneList_her2[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "\n",
    "# Normal data\n",
    "rowList_norm = list()\n",
    "for i in range(0, len(norm_exp)):\n",
    "    if np.sum(norm_exp[i,:]) == 0.0:\n",
    "        rowList_norm.append(i)\n",
    "\n",
    "norm_exp2 = np.zeros(((len(norm_exp)-len(rowList_norm)), len(norm_samples))) # matrix without zero rows\n",
    "print(norm_exp2.shape)\n",
    "geneList_norm2 = list()\n",
    "\n",
    "print(\"OLD \", len(norm_exp))\n",
    "print(\"ZEROS \", len(rowList_norm))\n",
    "print(\"NEW \", len(norm_exp2))\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(norm_exp)):\n",
    "    if j in rowList_norm:\n",
    "        j+=1\n",
    "    else:\n",
    "        norm_exp2[r,:] = norm_exp[j,:]\n",
    "        geneList_norm2.append(geneList_norm[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "print(len(geneList_norm2))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## log(TPM + 1) normalization\n",
    "# TPMi = (FPKMi / sum(FPKMj)) * 10^6, where i is the gene and j is the sample\n",
    "# logTPM = log(TPM + 1)\n",
    "\n",
    "# Luminal A\n",
    "for j in range(0, len(lumA_samples)):\n",
    "    sumPat = np.sum(lumA_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(lumA_exp2)):\n",
    "        lumA_exp2[i,j] = np.log2((lumA_exp2[i,j] / sumPat) * (10**6) + 1)\n",
    "        \n",
    "# Luminal B\n",
    "for j in range(0, len(lumB_samples)):\n",
    "    sumPat = np.sum(lumB_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(lumB_exp2)):\n",
    "        lumB_exp2[i,j] = np.log2((lumB_exp2[i,j] / sumPat) * (10**6) + 1)\n",
    "        \n",
    "# Basal\n",
    "for j in range(0, len(basal_samples)):\n",
    "    sumPat = np.sum(basal_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(basal_exp2)):\n",
    "        basal_exp2[i,j] = np.log2((basal_exp2[i,j] / sumPat) * (10**6) + 1)\n",
    "        \n",
    "# HER2\n",
    "for j in range(0, len(her2_samples)):\n",
    "    sumPat = np.sum(her2_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(her2_exp2)):\n",
    "        her2_exp2[i,j] = np.log2((her2_exp2[i,j] / sumPat) * (10**6) + 1)\n",
    "        \n",
    "# Normal\n",
    "for j in range(0, len(norm_samples)):\n",
    "    sumPat = np.sum(norm_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(norm_exp2)):\n",
    "        norm_exp2[i,j] = np.log2((norm_exp2[i,j] / sumPat) * (10**6) + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-24-653b83eb82c9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m             \u001b[0;31m# gets motif data for Luminal B\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 27\u001b[0;31m             \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgeneList_lumB2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     28\u001b[0m                 \u001b[0mmotifTF_lumB\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m                 \u001b[0mmotifGenes_lumB\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "## Extracts TFBM information\n",
    "# keeps track of which TF and genes have relations as per TFBM file\n",
    "\n",
    "motifPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/motif.txt' # original motif file\n",
    "\n",
    "motifTF_lumA = list() # list of TFs interacting to genes that are expressed\n",
    "motifGenes_lumA = list() # list of genes in expression file with TFs\n",
    "motifTF_lumB = list()\n",
    "motifGenes_lumB = list()\n",
    "motifTF_basal = list()\n",
    "motifGenes_basal = list()\n",
    "motifTF_her2 = list()\n",
    "motifGenes_her2 = list()\n",
    "motifTF_norm = list()\n",
    "motifGenes_norm = list()\n",
    "\n",
    "with open(motifPath, \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        if line.strip().split(\"\\t\")[2] == \"1\":\n",
    "            \n",
    "            # gets motif data for Luminal A\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_lumA2:\n",
    "                motifTF_lumA.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_lumA.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "            # gets motif data for Luminal B\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_lumB2:\n",
    "                motifTF_lumB.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_lumB.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "            # gets motif data for Basal\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_basal:\n",
    "                motifTF_basal.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_basal.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "            # gets motif data for HER2\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_her2:\n",
    "                motifTF_her2.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_her2.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "            # gets motif data for Normal\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_norm:\n",
    "                motifTF_norm.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_norm.append(line.strip().split(\"\\t\")[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(motifTF_lumA))\n",
    "print(len(motifGenes_lumA))\n",
    "\n",
    "print(len(motifTF_lumB))\n",
    "print(len(motifGenes_lumB))\n",
    "\n",
    "print(len(motifTF_basal))\n",
    "print(len(motifGenes_basal))\n",
    "\n",
    "print(len(motifTF_her2))\n",
    "print(len(motifGenes_her2))\n",
    "\n",
    "print(len(motifTF_norm))\n",
    "print(len(motifGenes_norm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Checks if there are genes expressed and not in the motif file\n",
    "# if not, then doesn't add to updated list of genes in the exp matrix\n",
    "# and makes note of it to remove later\n",
    "\n",
    "# Luminal A\n",
    "removeGene_lumA = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_lumA2)):\n",
    "    if geneList_lumA2[i] not in motifGenes_lumA:\n",
    "        removeGene_lumA.append(i)\n",
    "        \n",
    "print(len(removeGene_lumA))\n",
    "\n",
    "## Luminal B\n",
    "removeGene_lumB = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_lumB2)):\n",
    "    if geneList_lumB2[i] not in motifGenes_lumB:\n",
    "        removeGene_lumB.append(i)\n",
    "        \n",
    "print(len(removeGene_lumB))\n",
    "\n",
    "# Basal\n",
    "removeGene_basal = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_basal2)):\n",
    "    if geneList_basal2[i] not in motifGenes_basal:\n",
    "        removeGene_basal.append(i)\n",
    "        \n",
    "print(len(removeGene_basal))\n",
    "\n",
    "# HER2\n",
    "removeGene_her2 = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_her2_2)):\n",
    "    if geneList_her2_2[i] not in motifGenes_her2:\n",
    "        removeGene_her2.append(i)\n",
    "        \n",
    "print(len(removeGene_her2))\n",
    "\n",
    "# Normal\n",
    "removeGene_norm = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_norm2)):\n",
    "    if geneList_norm2[i] not in motifGenes_norm:\n",
    "        removeGene_norm.append(i)\n",
    "        \n",
    "print(len(removeGene_norm))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Removes the rows and genes that are not in the motif file\n",
    "\n",
    "# Luminal A\n",
    "lumA_exp3 = np.zeros((len(lumA_exp2)-len(removeGene_lumA), len(lumA_samples)))\n",
    "geneList_lumA3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(lumA_exp2)):\n",
    "    if i not in removeGene_lumA:\n",
    "        lumA_exp3[r,:] = lumA_exp2[i,:]\n",
    "        geneList_lumA3.append(geneList_lumA2[i])\n",
    "        r+=1\n",
    "\n",
    "# Luminal B\n",
    "lumB_exp3 = np.zeros((len(lumB_exp2)-len(removeGene_lumB), len(lumB_samples)))\n",
    "geneList_lumB3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(lumB_exp2)):\n",
    "    if i not in removeGene_lumB:\n",
    "        lumB_exp3[r,:] = lumB_exp2[i,:]\n",
    "        geneList_lumB3.append(geneList_lumB2[i])\n",
    "        r+=1\n",
    "\n",
    "# Basal\n",
    "basal_exp3 = np.zeros((len(basal_exp2)-len(removeGene_basal), len(basal_samples)))\n",
    "geneList_basal3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(basal_exp2)):\n",
    "    if i not in removeGene_basal:\n",
    "        basal_exp3[r,:] = basal_exp2[i,:]\n",
    "        geneList_basal3.append(geneList_basal2[i])\n",
    "        r+=1\n",
    "        \n",
    "# HER2\n",
    "her2_exp3 = np.zeros((len(her2_exp2)-len(removeGene_her2), len(her2_samples)))\n",
    "geneList_her2_3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(her2_exp2)):\n",
    "    if i not in removeGene_her2:\n",
    "        her2_exp3[r,:] = her2_exp2[i,:]\n",
    "        geneList_her2_3.append(geneList_her2_2[i])\n",
    "        r+=1\n",
    "        \n",
    "# Normal\n",
    "norm_exp3 = np.zeros((len(norm_exp2)-len(removeGene_norm), len(norm_samples)))\n",
    "geneList_norm3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(norm_exp2)):\n",
    "    if i not in removeGene_norm:\n",
    "        norm_exp3[r,:] = norm_exp2[i,:]\n",
    "        geneList_norm3.append(geneList_norm2[i])\n",
    "        r+=1\n",
    "        \n",
    "print(norm_exp3.shape)\n",
    "print(len(geneList_norm3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Removes genes the are not expressed but in motif file\n",
    "\n",
    "# gets all indeces of an element(gene) in a list(motifGenes)\n",
    "def getIndeces(listOfElements, element):\n",
    "    indexPosList = []\n",
    "    indexPos = 0\n",
    "    while True:\n",
    "        try:\n",
    "            # Search for item in list from indexPos to the end of list\n",
    "            indexPos = listOfElements.index(element, indexPos)\n",
    "            # Add the index position in list\n",
    "            indexPosList.append(indexPos)\n",
    "            indexPos += 1\n",
    "        except ValueError as e:\n",
    "            break\n",
    " \n",
    "    return indexPosList\n",
    "\n",
    "# Luminal A\n",
    "if set(motifGenes_lumA) != set(geneList_lumA3):\n",
    "    print(len(set(motifGenes_lumA)))\n",
    "    print(len(set(geneList_lumA3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_lumA) - set(geneList_lumA3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_lumA, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_lumA.remove(motifGenes_lumA[geneInds[i] - i])\n",
    "            motifTF_lumA.remove(motifTF_lumA[geneInds[i] - i])\n",
    "\n",
    "# Luminal B\n",
    "if set(motifGenes_lumB) != set(geneList_lumB3):\n",
    "    print(len(set(motifGenes_lumB)))\n",
    "    print(len(set(geneList_lumB3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_lumB) - set(geneList_lumB3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_lumB, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_lumB.remove(motifGenes_lumB[geneInds[i] - i])\n",
    "            motifTF_lumB.remove(motifTF_lumB[geneInds[i] - i])\n",
    "\n",
    "# Basal\n",
    "if set(motifGenes_basal) != set(geneList_basal3):\n",
    "    print(len(set(motifGenes_basal)))\n",
    "    print(len(set(geneList_basal3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_basal) - set(geneList_basal3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_basal, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_basal.remove(motifGenes_basal[geneInds[i] - i])\n",
    "            motifTF_basal.remove(motifTF_basal[geneInds[i] - i])\n",
    "            \n",
    "print(set(motifGenes_basal)== set(geneList_basal3))\n",
    "\n",
    "# HER2\n",
    "if set(motifGenes_her2) != set(geneList_her2_3):\n",
    "    print(len(set(motifGenes_her2)))\n",
    "    print(len(set(geneList_her2_3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_her2) - set(geneList_her2_3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_her2, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_her2.remove(motifGenes_her2[geneInds[i] - i])\n",
    "            motifTF_her2.remove(motifTF_her2[geneInds[i] - i])\n",
    "            \n",
    "print(set(motifGenes_her2)== set(geneList_her2_3))\n",
    "\n",
    "# Normal\n",
    "if set(motifGenes_norm) != set(geneList_norm3):\n",
    "    print(len(set(motifGenes_norm)))\n",
    "    print(len(set(geneList_norm3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_norm) - set(geneList_norm3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_norm, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_norm.remove(motifGenes_norm[geneInds[i] - i])\n",
    "            motifTF_norm.remove(motifTF_norm[geneInds[i] - i])\n",
    "            \n",
    "print(set(motifGenes_norm)== set(geneList_norm3))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(motifTF_basal))\n",
    "print(len(motifGenes_basal))\n",
    "\n",
    "print(len(motifTF_her2))\n",
    "print(len(motifGenes_her2))\n",
    "\n",
    "print(len(motifTF_norm))\n",
    "print(len(motifGenes_norm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove gene rows from HTSeq that are not in the network list\n",
    "\n",
    "# all_exp_HTSeq is geneList_HTSeq x indivList_HTSeq2\n",
    "\n",
    "#print(len(all_exp_HTSeq))\n",
    "#print(len(geneList_HTSeq))\n",
    "#print(len(indivList_HTSeq2))\n",
    "\n",
    "network_genes = list(set(geneList_lumA3 + geneList_lumB3 + geneList_basal3 + geneList_her2_3 + geneList_norm3))\n",
    "print(len(network_genes))\n",
    "\n",
    "#print(set(geneList_HTSeq))\n",
    "\n",
    "all_exp_HTSeq2 = np.zeros((len(network_genes), len(indivList_HTSeq2)))\n",
    "geneList_HTSeq2 = list()\n",
    "\n",
    "r = 0 # row index of all_exp_HTSeq2\n",
    "for i in range(0,len(geneList_HTSeq)):\n",
    "    if geneList_HTSeq[i] in network_genes:\n",
    "        all_exp_HTSeq2[r,:] = all_exp_HTSeq[i,:]\n",
    "        geneList_HTSeq2.append(geneList_HTSeq[i])\n",
    "        r+=1\n",
    "    \n",
    "print(len(geneList_HTSeq2))   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## makes the HTSeq file for the differential expression analysis\n",
    "\n",
    "## Makes the expression file\n",
    "\n",
    "PANDAexpPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/all_exp_HTSeq2.txt' # change name of new\n",
    "geneExpFile = open(PANDAexpPath,\"w+\")\n",
    "\n",
    "large_string = match_barcodes_HTSeq[0]\n",
    "for f in range(1, len(match_barcodes_HTSeq)):\n",
    "    large_string = large_string + \"\\t\" + match_barcodes_HTSeq[f]\n",
    "    \n",
    "newLine = large_string + \"\\n\"\n",
    "geneExpFile.write(newLine)\n",
    "\n",
    "for i in range(0, len(geneList_HTSeq2)):\n",
    "    newLine = geneList_HTSeq2[i]\n",
    "    \n",
    "    for j in range(0, len(file_id_HTSeq)):\n",
    "        newLine = newLine + \"\\t\" + str(all_exp_HTSeq2[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Reads in the ppi data and extracts the proteins \n",
    "# About 4 hrs to run\n",
    "\n",
    "ppiPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/ppi2015_freeze.txt'\n",
    "\n",
    "protein1_lumA = list()\n",
    "protein2_lumA = list()\n",
    "protein1_lumB = list()\n",
    "protein2_lumB = list()\n",
    "protein1_basal = list()\n",
    "protein2_basal = list()\n",
    "protein1_her2 = list()\n",
    "protein2_her2 = list()\n",
    "protein1_norm = list()\n",
    "protein2_norm = list()\n",
    "\n",
    "with open(ppiPath, \"r\") as file:\n",
    "    next(file) #skips the first line since that is just a header\n",
    "    for line in file.readlines():\n",
    "        if (line.strip().split(\"\\t\")[0] or line.strip().split(\"\\t\")[1]) in motifTF_lumA:\n",
    "            protein1_lumA.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_lumA.append(line.strip().split(\"\\t\")[1])\n",
    "        \n",
    "        if (line.strip().split(\"\\t\")[0] or line.strip().split(\"\\t\")[1]) in motifTF_lumB:\n",
    "            protein1_lumB.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_lumB.append(line.strip().split(\"\\t\")[1])\n",
    "            \n",
    "        if (line.strip().split(\"\\t\")[0] or line.strip().split(\"\\t\")[1]) in motifTF_basal:\n",
    "            protein1_basal.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_basal.append(line.strip().split(\"\\t\")[1])\n",
    "            \n",
    "        if (line.strip().split(\"\\t\")[0] or line.strip().split(\"\\t\")[1]) in motifTF_her2:\n",
    "            protein1_her2.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_her2.append(line.strip().split(\"\\t\")[1])\n",
    "            \n",
    "        if (line.strip().split(\"\\t\")[0] or line.strip().split(\"\\t\")[1]) in motifTF_norm:\n",
    "            protein1_norm.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_norm.append(line.strip().split(\"\\t\")[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Same!!\n",
      "Same!!\n",
      "Same!!\n",
      "Same!!\n",
      "Same!!\n"
     ]
    }
   ],
   "source": [
    "# Check to see if the two protein lists have the same proteins or not\n",
    "\n",
    "if(set(protein1_lumA) == set(protein2_lumA)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "if(set(protein1_lumB) == set(protein2_lumB)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "if(set(protein1_basal) == set(protein2_basal)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "if(set(protein1_her2) == set(protein2_her2)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "if(set(protein1_norm) == set(protein2_norm)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "# all proteins sets contain the same proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes the expression filess\n",
    "\n",
    "# Luminal A\n",
    "PANDAexpPath_lumA = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_lumA.txt'\n",
    "geneExpFile = open(PANDAexpPath_lumA,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_lumA3)):\n",
    "    newLine = geneList_lumA3[i]\n",
    "    \n",
    "    for j in range(0, len(lumA_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(lumA_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()\n",
    "\n",
    "# Luminal B\n",
    "PANDAexpPath_lumB = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_lumB.txt'\n",
    "geneExpFile = open(PANDAexpPath_lumB,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_lumB3)):\n",
    "    newLine = geneList_lumB3[i]\n",
    "    \n",
    "    for j in range(0, len(lumB_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(lumB_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()\n",
    "\n",
    "# Basal\n",
    "PANDAexpPath_basal = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_basal.txt'\n",
    "geneExpFile = open(PANDAexpPath_basal,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_basal3)):\n",
    "    newLine = geneList_basal3[i]\n",
    "    \n",
    "    for j in range(0, len(basal_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(basal_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()\n",
    "\n",
    "# HER2\n",
    "PANDAexpPath_her2 = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_her2.txt'\n",
    "geneExpFile = open(PANDAexpPath_her2,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_her2_3)):\n",
    "    newLine = geneList_her2_3[i]\n",
    "    \n",
    "    for j in range(0, len(her2_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(her2_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()\n",
    "\n",
    "# Normal\n",
    "PANDAexpPath_norm = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_norm.txt'\n",
    "geneExpFile = open(PANDAexpPath_norm,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_norm3)):\n",
    "    newLine = geneList_norm3[i]\n",
    "    \n",
    "    for j in range(0, len(norm_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(norm_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes the motif files\n",
    "\n",
    "# Luminal A\n",
    "PANDAmotifPath_lumA = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_lumA.txt'\n",
    "motifFile = open(PANDAmotifPath_lumA, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_lumA)):\n",
    "    newLine = motifTF_lumA[i] + \"\\t\" + motifGenes_lumA[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()\n",
    "\n",
    "# Luminal B\n",
    "PANDAmotifPath_lumB = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_lumB.txt'\n",
    "motifFile = open(PANDAmotifPath_lumB, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_lumB)):\n",
    "    newLine = motifTF_lumB[i] + \"\\t\" + motifGenes_lumB[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()\n",
    "\n",
    "# Basal\n",
    "PANDAmotifPath_basal = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_basal.txt'\n",
    "motifFile = open(PANDAmotifPath_basal, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_basal)):\n",
    "    newLine = motifTF_basal[i] + \"\\t\" + motifGenes_basal[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()\n",
    "\n",
    "# HER2\n",
    "PANDAmotifPath_her2 = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_her2.txt'\n",
    "motifFile = open(PANDAmotifPath_her2, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_her2)):\n",
    "    newLine = motifTF_her2[i] + \"\\t\" + motifGenes_her2[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()\n",
    "\n",
    "# Normal\n",
    "PANDAmotifPath_norm = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_norm.txt'\n",
    "motifFile = open(PANDAmotifPath_norm, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_norm)):\n",
    "    newLine = motifTF_norm[i] + \"\\t\" + motifGenes_norm[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes the ppi files\n",
    "\n",
    "# Luminal A\n",
    "PANDAppiPath_lumA = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_lumA.txt'\n",
    "ppiFile = open(PANDAppiPath_lumA, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_lumA)):\n",
    "    newLine = protein1_lumA[i] + \"\\t\" + protein2_lumA[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()\n",
    "\n",
    "# Luminal B\n",
    "PANDAppiPath_lumB = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_lumB.txt'\n",
    "ppiFile = open(PANDAppiPath_lumB, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_lumB)):\n",
    "    newLine = protein1_lumB[i] + \"\\t\" + protein2_lumB[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()\n",
    "\n",
    "# Basal\n",
    "PANDAppiPath_basal = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_basal.txt'\n",
    "ppiFile = open(PANDAppiPath_basal, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_basal)):\n",
    "    newLine = protein1_basal[i] + \"\\t\" + protein2_basal[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()\n",
    "\n",
    "# HER2\n",
    "PANDAppiPath_her2 = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_her2.txt'\n",
    "ppiFile = open(PANDAppiPath_her2, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_her2)):\n",
    "    newLine = protein1_her2[i] + \"\\t\" + protein2_her2[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()\n",
    "\n",
    "# Normal\n",
    "PANDAppiPath_norm = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_norm.txt'\n",
    "ppiFile = open(PANDAppiPath_norm, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_norm)):\n",
    "    newLine = protein1_norm[i] + \"\\t\" + protein2_norm[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "\n",
    "cd Desktop/BreCaNet/Code/pypanda\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_lumA.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_lumA.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_lumA.txt -o output_panda_lumA.txt\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_lumB.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_lumB.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_lumB.txt -o output_panda_lumB.txt\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_basal.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_basal.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_basal.txt -o output_panda_basal.txt\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_her2.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_her2.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_her2.txt -o output_panda_her2.txt\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_norm.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_norm.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_norm.txt -o output_panda_norm.txt\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
