{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BreCaNet\n",
    "\n",
    "This Jupyter notebook goes through the steps to process the full data sets before running PANDA\n",
    "\n",
    "##### Gene Expression Data\n",
    "\n",
    "Came from the TCGA database: \n",
    "Settings for data: *Cases tab*\n",
    "    Primary site -> breast,\n",
    "    Disease type -> ductal and lobular neoplasms,\n",
    "    Gender -> female,\n",
    "    Race -> White\n",
    " \n",
    " *Files tab*\n",
    "    Experimental Strategy -> RNA-seq,\n",
    "    Workflow -> HTSeq-FPKM,\n",
    "    Project -> TCGA-BRCA\n",
    "    \n",
    "Click on the Download button and select Cart\n",
    "\n",
    "##### Clinical Data with IHC Receptor Status\n",
    "\n",
    "Using the same settings as for gathering the expression data, except instead of downloading the cart, download the manifest data. The manifest file is used with the supplemental R code and necessary to retrieve the clinical IHC information.\n",
    "\n",
    "Use the same settings in the TCGA repository on the Cases tab.\n",
    "\n",
    "*Files tab:*\n",
    "Data Category -> Clinical\n",
    "\n",
    "Data Format -> BCR Biotab\n",
    "\n",
    "Add to cart \n",
    "\n",
    "Download the nationawidechildrens.org_clinical_patient_brca.txt file\n",
    "\n",
    "##### TF Binding Motif Data\n",
    "\n",
    "GTex from Camila\n",
    "\n",
    "##### Protien-Protein Interaction (PPI) Data\n",
    "\n",
    "GTex from Camila\n",
    "\n",
    "##### PANDA algorithm\n",
    "\n",
    "https://netzoo.github.io"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "In case there is a memory issue\n",
    "$ jupyter notebook --NotebookApp.iopub_data_rate_limit=10000000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Packages\n",
    "\n",
    "import gzip\n",
    "import numpy as np\n",
    "import os\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# User's paths for all files\n",
    "\n",
    "# gene expression paths\n",
    "dataPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/gdc_download_FPKM_832' #zipped file path\n",
    "dataNewFolder = '/Users/ursulawidocki/Desktop/BreCaNet/Data/processedTCGA_832' # folder for unzipped\n",
    "PANDAexpPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_832.txt' # final exp file\n",
    "\n",
    "# motif data paths\n",
    "motifPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/motif.txt' # original motif file\n",
    "motifNew = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/motifDataNew.txt'\n",
    "PANDAmotifPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs.txt' # file motif file\n",
    "\n",
    "# ppi data paths\n",
    "ppiPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/ppi2015_freeze.txt' # initial ppi path\n",
    "PANDAppiPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi/txt' # file ppi file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unzips the expression data dn creates a folder (processedTCGA) of the unzipped files \n",
    "# Gets the codes of the patient folders and expression files\n",
    "\n",
    "folderIDs = list() # codes of the patient IDs on the folder\n",
    "indivList = list() # codes of the patients' expression file\n",
    "\n",
    "for fileName in os.listdir(dataPath):\n",
    "    if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "        \n",
    "    else:\n",
    "        folderIDs.append(fileName)\n",
    "        \n",
    "    newPath = dataPath + '/' + fileName\n",
    "    \n",
    "    # goes into the individual folder\n",
    "    for indivName in os.listdir(newPath):\n",
    "        if (indivName == '.DS_Store') or (indivName == 'MANIFEST.txt') or (indivName == 'annotations.txt'):\n",
    "            continue\n",
    "        filePath = newPath + '/' + indivName\n",
    "        \n",
    "        fileEnding = indivName[-3] + indivName[-2] + indivName[-1]\n",
    "        if(fileEnding == '.gz'):\n",
    "            with gzip.open(filePath, 'rb') as f_in:\n",
    "                newFilePath = dataNewFolder + '/' + indivName[0:-3]\n",
    "                indivList.append(indivName[0:-3])\n",
    "                \n",
    "                with open(newFilePath, 'wb') as f_out:\n",
    "                    shutil.copyfileobj(f_in, f_out)\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# makes the folderIDs and indivList into files\n",
    "\n",
    "IDPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/IDList_FPKM.txt'\n",
    "folderIDPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/foldIDList_FPKM.txt'\n",
    "\n",
    "IDFile = open(IDPath, \"w+\")\n",
    "folderFile = open(folderIDPath, \"w+\")\n",
    "\n",
    "for i in range(0, 832):\n",
    "    newLineIDs = indivList[i][0:-9] + \"\\n\"\n",
    "    newLineFolder = folderIDs[i] + \"\\n\"\n",
    "    \n",
    "    IDFile.write(newLineIDs)\n",
    "    folderFile.write(newLineFolder)\n",
    "    \n",
    "IDFile.close()\n",
    "folderFile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "60483\n"
     ]
    }
   ],
   "source": [
    "# gets list of genes that are in the file\n",
    "\n",
    "geneList = list() # list of all genes that are noted in the expression files\n",
    "\n",
    "with open(dataNewFolder + '/' + indivList[0], \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        potentialGene = line.strip().split(\"\\t\")[0]\n",
    "        \n",
    "        if potentialGene[0:3] == \"ENS\":\n",
    "            newGeneName = potentialGene.strip().split(\".\")[0]\n",
    "            geneList.append(newGeneName)\n",
    "print(len(geneList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60483, 832)\n"
     ]
    }
   ],
   "source": [
    "# makes a matrix where patients are columns and genes are rows\n",
    "\n",
    "allTesting = np.zeros((len(geneList), len(indivList))) # matrix with all info from files\n",
    "j = 0 # index of the column (sample)\n",
    "i = 0 # index of the row (gene)\n",
    "print(allTesting.shape) # just to make sure dimensions are correct\n",
    "\n",
    "for fileName in os.listdir(dataNewFolder):\n",
    "    if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "        \n",
    "    if fileName in indivList:\n",
    "        temp_folder = folderIDs[indivList.index(fileName)]\n",
    "        readFile = dataNewFolder + '/' + fileName\n",
    "    \n",
    "        with open(readFile, \"r\") as file:\n",
    "            for line in file.readlines():\n",
    "                allTesting[i, j] = float(line.strip().split(\"\\t\")[1])\n",
    "                i+=1\n",
    "                \n",
    "            i = 0\n",
    "            if j < len(indivList):\n",
    "                j+=1\n",
    "            \n",
    "#for fileName in os.listdir(dataNewFolder):\n",
    "    #if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "    \n",
    "    #if fileName in indivList:\n",
    "        \n",
    "        #temp_folder = folderIDs[indivList.index(fileName)]\n",
    "        #readFile = dataNewFolder + '/' + fileName\n",
    "            \n",
    "        # adds to luminal A expression matrix\n",
    "        #if temp_folder in lumA_samples:\n",
    "            \n",
    "            #lumA_i = 0\n",
    "            #with open(readFile, \"r\") as file:\n",
    "                #for line in file.readlines():\n",
    "                    #lumA_exp[lumA_i, lumA_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    #lumA_i += 1\n",
    "                    \n",
    "                #lumA_j += 1"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## makes a file of the unfiltered and unnormalized data \n",
    "\n",
    "# converts ENS to gene symbol\n",
    "#convertFilePath = \"/Users/ursulawidocki/Desktop/BreCaNet/Data/geneNames.txt\"\n",
    "\n",
    "# gets Entrez IDs of genes \n",
    "#geneList_sym = list() # converted names of complete ENS list\n",
    "#with open(convertFilePath, \"r\") as file:\n",
    "    #next(file)\n",
    "    #for line in file.readlines():\n",
    "        #print(line.strip().split(\"\\t\")[0], line.strip().split(\"\\t\")[1])\n",
    "        #ensID = line.strip().split(\"\\t\")[0]\n",
    "        #if ensID in geneList:\n",
    "            #geneList[geneList.index(ensID)] = line.strip().split(\"\\t\")[1]\n",
    "\n",
    "#print(geneList[0:5])\n",
    "\n",
    "# makes the expression file\n",
    "\n",
    "PANDAexpPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_FPKM_unfilt.txt'\n",
    "geneExpFile = open(PANDAexpPath,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_sym)):\n",
    "    newLine = geneList_sym[i]\n",
    "    \n",
    "    for j in range(0, len(indivList)):\n",
    "        newLine = newLine + \"\\t\" + str(allTesting[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OLD  60483\n",
      "ZEROS  2393\n",
      "NEW  58090\n",
      "58090\n"
     ]
    }
   ],
   "source": [
    "## makes a matrix without zero rows and makes a list of only expressed genes\n",
    "# rows of zeros cause errors in PANDA\n",
    "\n",
    "# Makes note of which rows are zero rows\n",
    "rowList = list()\n",
    "for i in range(0, len(allTesting)):\n",
    "    if np.sum(allTesting[i,:]) == 0.0:\n",
    "        rowList.append(i)\n",
    "        \n",
    "allTesting2 = np.zeros(((len(allTesting)-len(rowList)), len(indivList))) # matrix without zero rows\n",
    "geneList2 = list() # list of only the expressed genes\n",
    "print(\"OLD \", len(allTesting))\n",
    "print(\"ZEROS \", len(rowList))\n",
    "print(\"NEW \", len(allTesting2))\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(allTesting)):\n",
    "    if j in rowList:\n",
    "        j+=1\n",
    "    else:\n",
    "        allTesting2[r,:] = allTesting[j,:]\n",
    "        geneList2.append(geneList[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "print(len(geneList2))\n",
    "del allTesting\n",
    "del geneList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "999999.999999999\n",
      "999999.9999999944\n",
      "1000000.0000000113\n",
      "999999.9999999946\n",
      "999999.999999997\n",
      "999999.9999999914\n",
      "999999.9999999987\n",
      "1000000.000000014\n",
      "1000000.0000000066\n",
      "1000000.0000000115\n",
      "1000000.0000000031\n",
      "1000000.0000000086\n",
      "1000000.0000000006\n",
      "999999.999999993\n",
      "1000000.0000000064\n",
      "999999.999999986\n",
      "1000000.0000000057\n",
      "999999.9999999962\n",
      "999999.9999999921\n",
      "999999.9999999952\n",
      "1000000.0000000118\n",
      "1000000.0000000079\n",
      "999999.9999999978\n",
      "1000000.0000000092\n",
      "999999.9999999881\n",
      "1000000.0000000007\n",
      "1000000.0000000114\n",
      "1000000.0000000028\n",
      "1000000.0000000104\n",
      "1000000.0000000044\n",
      "999999.9999999925\n",
      "1000000.0000000009\n",
      "999999.9999999919\n",
      "1000000.0000000108\n",
      "1000000.000000009\n",
      "999999.9999999963\n",
      "1000000.0000000088\n",
      "999999.9999999979\n",
      "1000000.0000000021\n",
      "999999.9999999695\n",
      "1000000.0000000088\n",
      "1000000.0000000128\n",
      "1000000.0000000029\n",
      "999999.999999991\n",
      "999999.9999999944\n",
      "999999.9999999976\n",
      "999999.999999991\n",
      "999999.9999999938\n",
      "1000000.0000000073\n",
      "1000000.0000000128\n",
      "999999.9999999832\n",
      "999999.999999995\n",
      "1000000.000000002\n",
      "1000000.000000002\n",
      "1000000.0000000116\n",
      "1000000.0000000068\n",
      "999999.9999999967\n",
      "1000000.0000000021\n",
      "1000000.0000000066\n",
      "1000000.0000000091\n",
      "1000000.0000000079\n",
      "1000000.0000000108\n",
      "1000000.0000000016\n",
      "999999.9999999903\n",
      "999999.9999999929\n",
      "999999.9999999845\n",
      "999999.9999999669\n",
      "1000000.0000000017\n",
      "1000000.0000000079\n",
      "1000000.0\n",
      "1000000.0000000012\n",
      "1000000.0000000016\n",
      "1000000.0000000073\n",
      "1000000.0000000095\n",
      "1000000.0000000107\n",
      "999999.9999999991\n",
      "1000000.0000000013\n",
      "999999.999999996\n",
      "1000000.0000000019\n",
      "1000000.0000000113\n",
      "999999.9999999924\n",
      "1000000.0000000055\n",
      "999999.9999999984\n",
      "1000000.0000000102\n",
      "999999.9999999889\n",
      "999999.9999999912\n",
      "999999.9999999938\n",
      "999999.9999999965\n",
      "1000000.0000000001\n",
      "1000000.000000003\n",
      "999999.9999999916\n",
      "1000000.0000000077\n",
      "999999.9999999971\n",
      "1000000.0000000023\n",
      "1000000.0000000199\n",
      "1000000.0\n",
      "1000000.000000002\n",
      "999999.9999999956\n",
      "1000000.0000000083\n",
      "1000000.0000000021\n",
      "999999.9999999919\n",
      "1000000.0000000047\n",
      "999999.999999995\n",
      "999999.9999999934\n",
      "1000000.0000000013\n",
      "999999.9999999924\n",
      "1000000.0000000019\n",
      "999999.9999999974\n",
      "999999.9999999938\n",
      "1000000.0000000084\n",
      "999999.9999999931\n",
      "1000000.0000000062\n",
      "1000000.0000000051\n",
      "1000000.0000000182\n",
      "1000000.000000003\n",
      "1000000.000000015\n",
      "1000000.0000000006\n",
      "1000000.000000001\n",
      "999999.9999999929\n",
      "999999.9999999971\n",
      "1000000.0000000056\n",
      "999999.9999999983\n",
      "1000000.0000000052\n",
      "1000000.0000000115\n",
      "999999.9999999874\n",
      "1000000.0000000062\n",
      "999999.9999999921\n",
      "999999.9999999965\n",
      "1000000.0000000101\n",
      "1000000.0000000009\n",
      "999999.9999999962\n",
      "999999.9999999945\n",
      "999999.9999999925\n",
      "1000000.0000000108\n",
      "1000000.0000000008\n",
      "1000000.0000000057\n",
      "999999.9999999962\n",
      "999999.9999999997\n",
      "999999.999999999\n",
      "1000000.0000000114\n",
      "1000000.0000000076\n",
      "1000000.0000000114\n",
      "999999.999999987\n",
      "999999.9999999969\n",
      "1000000.0000000107\n",
      "1000000.0000000057\n",
      "1000000.0000000057\n",
      "1000000.0000000021\n",
      "1000000.0000000118\n",
      "1000000.0000000107\n",
      "999999.9999999964\n",
      "1000000.0000000173\n",
      "1000000.0000000137\n",
      "1000000.0000000112\n",
      "999999.999999998\n",
      "1000000.0000000048\n",
      "1000000.0000000073\n",
      "1000000.0000000121\n",
      "999999.9999999951\n",
      "999999.9999999985\n",
      "1000000.0000000126\n",
      "999999.9999999987\n",
      "1000000.0000000071\n",
      "1000000.0000000017\n",
      "1000000.0000000021\n",
      "999999.9999999878\n",
      "999999.9999999878\n",
      "1000000.000000007\n",
      "999999.9999999877\n",
      "1000000.0000000016\n",
      "1000000.0000000129\n",
      "1000000.0000000075\n",
      "999999.999999987\n",
      "999999.9999999958\n",
      "999999.999999998\n",
      "999999.9999999937\n",
      "999999.999999998\n",
      "1000000.0\n",
      "1000000.0000000227\n",
      "1000000.0000000023\n",
      "999999.9999999995\n",
      "999999.9999999976\n",
      "999999.9999999944\n",
      "1000000.0000000042\n",
      "999999.9999999847\n",
      "1000000.0000000016\n",
      "1000000.0000000042\n",
      "1000000.0000000026\n",
      "1000000.0000000024\n",
      "999999.9999999775\n",
      "999999.9999999902\n",
      "999999.9999999952\n",
      "1000000.0000000154\n",
      "1000000.0000000115\n",
      "1000000.0000000217\n",
      "1000000.0000000043\n",
      "999999.9999999979\n",
      "1000000.0000000014\n",
      "1000000.0000000024\n",
      "999999.9999999969\n",
      "1000000.0000000121\n",
      "1000000.000000006\n",
      "1000000.0000000033\n",
      "1000000.0000000005\n",
      "999999.9999999981\n",
      "1000000.0000000148\n",
      "1000000.0000000081\n",
      "1000000.0000000048\n",
      "1000000.0000000122\n",
      "999999.9999999983\n",
      "999999.9999999892\n",
      "999999.9999999896\n",
      "1000000.0000000022\n",
      "999999.9999999871\n",
      "999999.9999999964\n",
      "1000000.0000000022\n",
      "999999.9999999964\n",
      "999999.9999999901\n",
      "1000000.0000000079\n",
      "1000000.0000000008\n",
      "1000000.0000000158\n",
      "1000000.000000005\n",
      "1000000.0000000134\n",
      "1000000.0000000088\n",
      "999999.9999999999\n",
      "999999.999999995\n",
      "999999.9999999899\n",
      "999999.9999999935\n",
      "999999.9999999903\n",
      "1000000.0000000051\n",
      "1000000.0000000156\n",
      "1000000.0000000199\n",
      "999999.9999999973\n",
      "1000000.0000000127\n",
      "1000000.0000000119\n",
      "1000000.0000000015\n",
      "1000000.0000000037\n",
      "999999.9999999957\n",
      "1000000.0000000065\n",
      "1000000.0000000002\n",
      "1000000.0000000023\n",
      "1000000.000000004\n",
      "1000000.0000000037\n",
      "1000000.0000000007\n",
      "999999.9999999988\n",
      "1000000.0000000013\n",
      "999999.9999999955\n",
      "999999.9999999917\n",
      "1000000.0000000084\n",
      "1000000.0000000084\n",
      "999999.9999999997\n",
      "1000000.0000000013\n",
      "999999.9999999861\n",
      "999999.9999999957\n",
      "999999.9999999864\n",
      "999999.9999999863\n",
      "1000000.000000003\n",
      "1000000.0000000186\n",
      "999999.9999999902\n",
      "999999.9999999854\n",
      "999999.9999999928\n",
      "1000000.0000000071\n",
      "1000000.0000000043\n",
      "999999.9999999995\n",
      "999999.999999987\n",
      "999999.9999999993\n",
      "999999.9999999983\n",
      "999999.9999999872\n",
      "1000000.0000000054\n",
      "1000000.0000000093\n",
      "999999.9999999925\n",
      "999999.9999999871\n",
      "1000000.0000000014\n",
      "1000000.0000000091\n",
      "1000000.0000000035\n",
      "999999.9999999899\n",
      "1000000.0000000033\n",
      "1000000.0000000056\n",
      "999999.9999999962\n",
      "999999.9999999945\n",
      "999999.9999999946\n",
      "999999.9999999836\n",
      "999999.999999989\n",
      "999999.9999999899\n",
      "999999.9999999972\n",
      "999999.9999999925\n",
      "1000000.000000008\n",
      "999999.9999999916\n",
      "1000000.0000000191\n",
      "1000000.0000000013\n",
      "1000000.0000000114\n",
      "1000000.0000000088\n",
      "999999.9999999949\n",
      "1000000.0000000057\n",
      "999999.9999999999\n",
      "999999.9999999846\n",
      "1000000.0000000123\n",
      "1000000.0000000075\n",
      "1000000.0000000049\n",
      "999999.9999999881\n",
      "999999.9999999994\n",
      "999999.9999999912\n",
      "1000000.0000000014\n",
      "1000000.0000000234\n",
      "999999.9999999942\n",
      "999999.9999999892\n",
      "999999.9999999955\n",
      "999999.9999999931\n",
      "1000000.0000000026\n",
      "999999.9999999942\n",
      "1000000.0000000065\n",
      "1000000.0000000019\n",
      "999999.9999999903\n",
      "999999.9999999952\n",
      "999999.9999999917\n",
      "999999.9999999948\n",
      "1000000.0000000034\n",
      "999999.9999999976\n",
      "1000000.0000000084\n",
      "999999.9999999906\n",
      "999999.9999999994\n",
      "1000000.0000000122\n",
      "999999.999999999\n",
      "1000000.0000000065\n",
      "999999.9999999978\n",
      "1000000.0000000057\n",
      "999999.9999999978\n",
      "1000000.0000000076\n",
      "1000000.0000000211\n",
      "1000000.0000000015\n",
      "1000000.000000002\n",
      "999999.9999999976\n",
      "1000000.000000017\n",
      "1000000.0000000012\n",
      "1000000.0000000037\n",
      "1000000.0000000029\n",
      "999999.9999999885\n",
      "999999.999999989\n",
      "999999.999999995\n",
      "1000000.0000000035\n",
      "999999.9999999956\n",
      "1000000.0000000023\n",
      "1000000.000000004\n",
      "999999.9999999822\n",
      "999999.999999993\n",
      "1000000.0000000021\n",
      "1000000.0000000158\n",
      "999999.9999999995\n",
      "1000000.0000000097\n",
      "999999.999999999\n",
      "1000000.0000000106\n",
      "999999.9999999979\n",
      "1000000.0000000186\n",
      "999999.9999999907\n",
      "999999.999999995\n",
      "1000000.0000000143\n",
      "1000000.0000000009\n",
      "999999.9999999991\n",
      "1000000.0000000051\n",
      "1000000.0000000211\n",
      "1000000.0000000064\n",
      "1000000.00000001\n",
      "1000000.0000000105\n",
      "1000000.0000000246\n",
      "1000000.0000000084\n",
      "999999.9999999882\n",
      "999999.9999999959\n",
      "999999.9999999851\n",
      "1000000.0000000068\n",
      "1000000.0000000071\n",
      "999999.9999999764\n",
      "1000000.0000000113\n",
      "1000000.0000000031\n",
      "999999.9999999998\n",
      "1000000.0000000141\n",
      "1000000.0000000105\n",
      "1000000.0000000087\n",
      "1000000.0000000026\n",
      "1000000.0000000024\n",
      "1000000.000000005\n",
      "1000000.0000000079\n",
      "1000000.0000000135\n",
      "999999.9999999931\n",
      "999999.9999999938\n",
      "1000000.0000000059\n",
      "1000000.0000000059\n",
      "1000000.0000000005\n",
      "1000000.0000000027\n",
      "999999.9999999986\n",
      "1000000.0000000068\n",
      "1000000.0000000071\n",
      "999999.9999999858\n",
      "1000000.0000000038\n",
      "999999.9999999985\n",
      "1000000.0000000013\n",
      "1000000.000000005\n",
      "1000000.0000000013\n",
      "999999.9999999956\n",
      "1000000.0000000069\n",
      "999999.9999999865\n",
      "1000000.0000000163\n",
      "1000000.0000000026\n",
      "999999.9999999955\n",
      "999999.9999999882\n",
      "1000000.0000000044\n",
      "1000000.0000000137\n",
      "1000000.0000000172\n",
      "1000000.0000000007\n",
      "999999.9999999929\n",
      "999999.9999999785\n",
      "1000000.0000000033\n",
      "999999.9999999969\n",
      "999999.9999999898\n",
      "1000000.0000000063\n",
      "1000000.0000000026\n",
      "999999.9999999859\n",
      "999999.9999999963\n",
      "1000000.0000000001\n",
      "1000000.000000003\n",
      "999999.9999999986\n",
      "1000000.0000000008\n",
      "1000000.0000000098\n",
      "999999.9999999804\n",
      "1000000.0000000022\n",
      "1000000.0000000045\n",
      "999999.9999999953\n",
      "1000000.0000000022\n",
      "999999.9999999969\n",
      "999999.9999999885\n",
      "1000000.000000003\n",
      "1000000.0000000139\n",
      "1000000.0000000193\n",
      "999999.9999999962\n",
      "1000000.0000000017\n",
      "999999.9999999979\n",
      "999999.9999999853\n",
      "1000000.0000000056\n",
      "999999.9999999972\n",
      "999999.9999999969\n",
      "999999.9999999997\n",
      "1000000.0000000115\n",
      "999999.9999999928\n",
      "999999.9999999935\n",
      "999999.9999999942\n",
      "1000000.0000000058\n",
      "999999.9999999816\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "999999.9999999801\n",
      "999999.9999999965\n",
      "1000000.0000000002\n",
      "999999.9999999991\n",
      "999999.9999999894\n",
      "1000000.0000000121\n",
      "999999.9999999905\n",
      "999999.9999999778\n",
      "1000000.0000000021\n",
      "1000000.000000001\n",
      "999999.9999999998\n",
      "999999.9999999991\n",
      "1000000.0000000073\n",
      "999999.999999996\n",
      "1000000.0000000033\n",
      "1000000.000000003\n",
      "1000000.000000017\n",
      "999999.9999999977\n",
      "1000000.0000000005\n",
      "999999.9999999868\n",
      "999999.9999999985\n",
      "999999.9999999978\n",
      "999999.9999999987\n",
      "1000000.0000000072\n",
      "1000000.0000000069\n",
      "999999.9999999983\n",
      "1000000.0000000009\n",
      "1000000.0000000008\n",
      "1000000.0000000023\n",
      "1000000.0000000017\n",
      "1000000.0000000143\n",
      "1000000.000000006\n",
      "999999.9999999934\n",
      "999999.9999999845\n",
      "999999.9999999951\n",
      "1000000.0000000059\n",
      "1000000.0000000001\n",
      "999999.9999999995\n",
      "999999.9999999831\n",
      "999999.9999999806\n",
      "1000000.0000000033\n",
      "1000000.0000000208\n",
      "999999.9999999978\n",
      "1000000.0000000014\n",
      "1000000.0000000111\n",
      "1000000.0000000052\n",
      "1000000.000000015\n",
      "1000000.0000000024\n",
      "1000000.0000000041\n",
      "999999.9999999879\n",
      "1000000.0000000116\n",
      "999999.9999999966\n",
      "1000000.0000000021\n",
      "1000000.0000000085\n",
      "999999.999999987\n",
      "999999.9999999857\n",
      "1000000.0000000083\n",
      "1000000.0000000081\n",
      "999999.9999999976\n",
      "999999.9999999997\n",
      "1000000.0000000023\n",
      "999999.9999999965\n",
      "1000000.0000000035\n",
      "999999.999999996\n",
      "999999.9999999977\n",
      "999999.9999999949\n",
      "999999.999999995\n",
      "999999.9999999977\n",
      "1000000.0000000077\n",
      "1000000.0000000066\n",
      "1000000.0000000002\n",
      "1000000.000000002\n",
      "999999.9999999997\n",
      "1000000.0000000072\n",
      "1000000.0000000048\n",
      "999999.9999999778\n",
      "1000000.0000000142\n",
      "1000000.0000000006\n",
      "999999.9999999912\n",
      "999999.9999999945\n",
      "1000000.0000000026\n",
      "1000000.0000000055\n",
      "1000000.0000000169\n",
      "1000000.000000008\n",
      "1000000.0000000027\n",
      "1000000.0000000036\n",
      "1000000.0000000015\n",
      "999999.9999999894\n",
      "999999.999999998\n",
      "1000000.0000000098\n",
      "999999.9999999892\n",
      "1000000.0000000068\n",
      "999999.9999999962\n",
      "1000000.00000001\n",
      "1000000.0000000132\n",
      "999999.9999999946\n",
      "999999.9999999986\n",
      "999999.9999999976\n",
      "999999.9999999997\n",
      "1000000.000000001\n",
      "999999.9999999992\n",
      "1000000.0000000222\n",
      "999999.9999999909\n",
      "1000000.0000000157\n",
      "999999.9999999881\n",
      "999999.9999999856\n",
      "1000000.0000000019\n",
      "1000000.000000003\n",
      "999999.999999998\n",
      "999999.9999999977\n",
      "999999.9999999854\n",
      "1000000.0000000199\n",
      "1000000.0000000083\n",
      "999999.9999999941\n",
      "1000000.0000000019\n",
      "999999.9999999864\n",
      "999999.9999999962\n",
      "999999.9999999984\n",
      "1000000.000000006\n",
      "1000000.000000001\n",
      "999999.9999999973\n",
      "1000000.0000000079\n",
      "1000000.0000000052\n",
      "999999.9999999966\n",
      "1000000.0000000175\n",
      "1000000.0000000008\n",
      "999999.9999999865\n",
      "999999.9999999967\n",
      "1000000.0000000038\n",
      "1000000.0000000051\n",
      "1000000.0000000069\n",
      "1000000.0000000065\n",
      "1000000.0000000051\n",
      "999999.999999992\n",
      "1000000.0000000038\n",
      "999999.9999999851\n",
      "1000000.0000000016\n",
      "1000000.000000005\n",
      "1000000.000000005\n",
      "999999.9999999935\n",
      "999999.999999987\n",
      "999999.9999999992\n",
      "1000000.0000000066\n",
      "1000000.000000015\n",
      "1000000.0000000052\n",
      "1000000.0000000042\n",
      "999999.9999999997\n",
      "999999.9999999809\n",
      "999999.9999999993\n",
      "1000000.000000014\n",
      "1000000.0000000044\n",
      "1000000.000000004\n",
      "1000000.0000000042\n",
      "1000000.0000000356\n",
      "999999.9999999984\n",
      "1000000.0000000003\n",
      "1000000.000000009\n",
      "1000000.0000000005\n",
      "999999.9999999927\n",
      "1000000.0000000016\n",
      "1000000.0000000002\n",
      "1000000.0000000099\n",
      "1000000.0000000026\n",
      "1000000.0000000044\n",
      "999999.9999999834\n",
      "1000000.0000000037\n",
      "1000000.0000000009\n",
      "1000000.0000000183\n",
      "1000000.000000005\n",
      "1000000.000000002\n",
      "999999.999999994\n",
      "999999.9999999922\n",
      "1000000.0000000022\n",
      "1000000.0000000091\n",
      "999999.9999999995\n",
      "1000000.0000000024\n",
      "999999.9999999965\n",
      "1000000.0000000027\n",
      "1000000.0000000054\n",
      "999999.9999999942\n",
      "1000000.0000000057\n",
      "999999.9999999984\n",
      "1000000.000000007\n",
      "1000000.0000000068\n",
      "1000000.0000000009\n",
      "1000000.0000000211\n",
      "1000000.0000000003\n",
      "1000000.0000000095\n",
      "1000000.0000000075\n",
      "1000000.0000000034\n",
      "999999.999999995\n",
      "999999.9999999936\n",
      "999999.9999999885\n",
      "1000000.0000000036\n",
      "999999.9999999866\n",
      "1000000.0000000048\n",
      "999999.9999999976\n",
      "999999.999999995\n",
      "1000000.0000000008\n",
      "1000000.0000000056\n",
      "1000000.0000000023\n",
      "999999.999999983\n",
      "999999.9999999995\n",
      "1000000.0000000001\n",
      "999999.9999999936\n",
      "999999.9999999969\n",
      "999999.9999999941\n",
      "1000000.0000000073\n",
      "1000000.0000000189\n",
      "1000000.0000000009\n",
      "999999.9999999951\n",
      "999999.9999999912\n",
      "999999.9999999969\n",
      "1000000.0000000022\n",
      "1000000.0000000099\n",
      "999999.9999999953\n",
      "999999.9999999919\n",
      "999999.9999999867\n",
      "999999.9999999967\n",
      "999999.999999982\n",
      "1000000.0000000062\n",
      "1000000.0\n",
      "999999.9999999997\n",
      "999999.9999999888\n",
      "999999.9999999877\n",
      "1000000.0000000038\n",
      "1000000.0000000111\n",
      "999999.9999999789\n",
      "1000000.0000000007\n",
      "1000000.0000000116\n",
      "999999.9999999931\n",
      "1000000.0000000054\n",
      "999999.9999999999\n",
      "1000000.0000000154\n",
      "999999.999999998\n",
      "999999.999999998\n",
      "999999.9999999966\n",
      "1000000.0000000048\n",
      "999999.9999999965\n",
      "1000000.0000000132\n",
      "999999.9999999944\n",
      "1000000.0000000055\n",
      "1000000.0000000048\n",
      "1000000.0000000076\n",
      "1000000.0000000088\n",
      "1000000.0000000037\n",
      "1000000.0000000123\n",
      "1000000.0000000099\n",
      "999999.999999997\n",
      "1000000.000000007\n",
      "999999.99999999\n",
      "1000000.000000008\n",
      "999999.9999999941\n",
      "999999.9999999984\n",
      "1000000.0000000132\n",
      "999999.9999999879\n",
      "1000000.0000000142\n",
      "1000000.0000000157\n",
      "1000000.0000000051\n",
      "1000000.0000000104\n",
      "1000000.0000000087\n",
      "1000000.0000000071\n",
      "999999.9999999978\n",
      "999999.9999999965\n",
      "999999.9999999983\n",
      "999999.9999999995\n",
      "999999.9999999957\n",
      "999999.9999999827\n",
      "999999.9999999978\n",
      "999999.9999999864\n",
      "999999.9999999894\n",
      "1000000.0000000072\n",
      "999999.9999999864\n",
      "1000000.0000000019\n",
      "1000000.0000000033\n",
      "999999.9999999985\n",
      "999999.9999999932\n",
      "1000000.000000009\n",
      "1000000.0000000161\n",
      "999999.9999999828\n",
      "1000000.0000000036\n",
      "999999.9999999985\n",
      "999999.9999999969\n",
      "999999.999999997\n",
      "1000000.0000000098\n",
      "999999.9999999831\n",
      "999999.9999999885\n",
      "999999.9999999971\n",
      "1000000.0000000045\n",
      "1000000.0000000254\n",
      "1000000.0000000042\n",
      "1000000.0000000094\n",
      "1000000.0000000143\n",
      "999999.9999999972\n",
      "1000000.0000000222\n",
      "999999.9999999893\n",
      "999999.9999999966\n",
      "1000000.000000003\n",
      "1000000.0000000122\n",
      "999999.9999999944\n",
      "999999.9999999983\n",
      "1000000.0000000146\n",
      "1000000.0000000058\n",
      "999999.999999993\n",
      "999999.9999999944\n",
      "1000000.0000000029\n",
      "999999.9999999995\n",
      "999999.9999999874\n",
      "999999.9999999902\n",
      "1000000.0000000085\n",
      "999999.9999999987\n",
      "999999.9999999958\n",
      "1000000.0000000002\n",
      "999999.9999999945\n",
      "999999.9999999965\n",
      "1000000.0000000056\n",
      "999999.999999991\n",
      "999999.999999993\n",
      "1000000.000000002\n",
      "1000000.000000007\n",
      "999999.9999999878\n",
      "1000000.000000016\n",
      "999999.9999999935\n",
      "999999.9999999997\n",
      "999999.9999999962\n",
      "999999.9999999923\n",
      "999999.9999999946\n",
      "999999.9999999949\n",
      "1000000.0000000255\n",
      "1000000.0000000041\n",
      "999999.9999999927\n",
      "1000000.0000000049\n",
      "999999.9999999983\n",
      "1000000.0000000015\n",
      "1000000.00000002\n",
      "999999.9999999921\n",
      "999999.9999999903\n",
      "999999.9999999913\n",
      "999999.9999999986\n",
      "999999.9999999844\n",
      "999999.9999999941\n",
      "999999.9999999992\n",
      "1000000.0000000001\n",
      "999999.9999999952\n",
      "999999.9999999986\n",
      "999999.9999999813\n",
      "999999.9999999991\n",
      "1000000.0000000062\n",
      "1000000.0000000108\n",
      "1000000.0000000212\n",
      "1000000.0000000081\n",
      "1000000.000000005\n",
      "1000000.0000000087\n",
      "1000000.0000000015\n",
      "999999.9999999908\n",
      "999999.9999999894\n",
      "1000000.0000000116\n",
      "1000000.0000000026\n",
      "1000000.000000005\n",
      "999999.9999999969\n",
      "1000000.0000000075\n",
      "1000000.0000000076\n",
      "1000000.0000000007\n",
      "1000000.0000000063\n",
      "1000000.0000000044\n",
      "999999.9999999873\n",
      "1000000.0000000073\n",
      "1000000.0000000098\n",
      "1000000.0000000037\n",
      "999999.99999999\n",
      "1000000.0000000024\n",
      "999999.9999999966\n",
      "999999.9999999892\n",
      "999999.9999999905\n",
      "1000000.0000000086\n",
      "999999.9999999934\n",
      "1000000.0000000021\n",
      "1000000.000000007\n",
      "999999.9999999873\n",
      "1000000.0000000001\n",
      "999999.9999999992\n",
      "1000000.000000002\n",
      "999999.9999999962\n",
      "999999.9999999952\n",
      "1000000.0000000042\n",
      "999999.9999999895\n"
     ]
    }
   ],
   "source": [
    "## TPM normalization\n",
    "# because this allows genes to be compared well across samples\n",
    "## TPMi = (FPKMi / sum(FPKMj)) * 10^6, where i is the gene and j is the sample\n",
    "\n",
    "for j in range(0, len(indivList)):\n",
    "    sumPat = np.sum(allTesting2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(allTesting2)):\n",
    "        allTesting2[i,j] = (allTesting2[i,j] / sumPat) * (10**6)\n",
    "        \n",
    "# makes sure that the rows add up to the same sum\n",
    "\n",
    "for j in range(0, len(allTesting2[0])):\n",
    "    sumPat = np.sum(allTesting2, axis = 0)[j]\n",
    "    print(sumPat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract TFBM information if the gene is a gene expressed\n",
    "\n",
    "motifTFs = list() # list of TFs with genes that are expressed\n",
    "motifGenes = list() # list of genes with TFs and in expression file\n",
    "\n",
    "with open(motifPath, \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        if line.strip().split(\"\\t\")[2] == \"1\":\n",
    "            if line.strip().split(\"\\t\")[1] in geneList2:\n",
    "                motifTFs.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes.append(line.strip().split(\"\\t\")[1])\n",
    "                "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "del folderIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1792776\n",
      "1792776\n"
     ]
    }
   ],
   "source": [
    "print(len(motifTFs))\n",
    "print(len(motifGenes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29216\n"
     ]
    }
   ],
   "source": [
    "# checks if there are genes expressed and not in the motif file\n",
    "# if not, then doesn't add to updated list of genes in the exp matrix\n",
    "\n",
    "rowList2 = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList2)):\n",
    "    if geneList2[i] not in motifGenes:\n",
    "        rowList2.append(i)\n",
    "        \n",
    "print(len(rowList2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28874, 832)\n",
      "28874\n"
     ]
    }
   ],
   "source": [
    "# removes the rows and genes that are not in the motif file\n",
    "\n",
    "allTesting3 = np.zeros((len(allTesting2)-len(rowList2), len(indivList)))\n",
    "geneList3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(allTesting2)):\n",
    "    if i not in rowList2:\n",
    "        allTesting3[r,:] = allTesting2[i,:]\n",
    "        geneList3.append(geneList2[i])\n",
    "        r+=1\n",
    "        \n",
    "print(allTesting3.shape)\n",
    "print(len(geneList3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to save memory\n",
    "del allTesting2\n",
    "del geneList2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reads in the ppi data and extracts the proteins \n",
    "\n",
    "ppiPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/ppi2015_freeze.txt'\n",
    "\n",
    "protein1 = list()\n",
    "protein2 = list()\n",
    "\n",
    "with open(ppiPath, \"r\") as file:\n",
    "    next(file) #skips the first line since that is just a header\n",
    "    for line in file.readlines():\n",
    "        if line.strip().split(\"\\t\")[0] in motifTFs:\n",
    "            protein1.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2.append(line.strip().split(\"\\t\")[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "84387"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add checks here\n",
    "len(protein1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "# convert ENS IDs to Entrez\n",
    "\n",
    "convertFilePath = \"/Users/ursulawidocki/Desktop/BreCaNet/Data/geneNames.txt\"\n",
    "\n",
    "# gets Entrez IDs of genes \n",
    "with open(convertFilePath, \"r\") as file:\n",
    "    next(file)\n",
    "    for line in file.readlines():\n",
    "        #print(line.strip().split(\"\\t\")[0], line.strip().split(\"\\t\")[1])\n",
    "        ensID = line.strip().split(\"\\t\")[0]\n",
    "        if ensID in geneList3:\n",
    "            geneList3[geneList3.index(ensID)] = line.strip().split(\"\\t\")[1]\n",
    "\n",
    "# locates duplicate genes\n",
    "print(len(geneList3) - len(set(geneList3))) # there are no duplicates"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# TPM normalization\n",
    "# because this allows genes to be compared well across samples\n",
    "## TPMi = (FPKMi / sum(FPKMj)) * 10^6, where i is the gene and j is the sample\n",
    "\n",
    "for j in range(0, len(indivList)):\n",
    "    sumPat = np.sum(allTesting3, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(allTesting3)):\n",
    "        allTesting3[i,j] = (allTesting3[i,j] / sumPat) * (10**6)\n",
    "        \n",
    "# makes sure that the rows add up to the same sum\n",
    "\n",
    "for j in range(0, len(allTesting3[0])):\n",
    "    sumPat = np.sum(allTesting3, axis = 0)[j]\n",
    "    print(sumPat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# makes the expression file\n",
    "\n",
    "PANDAexpPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp.txt'\n",
    "geneExpFile = open(PANDAexpPath,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList3)):\n",
    "    newLine = geneList3[i]\n",
    "    \n",
    "    for j in range(0, len(indivList)):\n",
    "        newLine = newLine + \"\\t\" + str(allTesting3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# makes the motif file\n",
    "\n",
    "PANDAmotifPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs.txt'\n",
    "motifFile = open(PANDAmotifPath, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTFs)):\n",
    "    newLine = motifTFs[i] + \"\\t\" + motifGenes[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# makes the ppi file\n",
    "\n",
    "PANDAppiPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi.txt'\n",
    "ppiFile = open(PANDAppiPath, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1)):\n",
    "    newLine = protein1[i] + \"\\t\" + protein2[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# makes a file of the Entrez IDs\n",
    "\n",
    "PANDAgeneListPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/finalGeneList.txt'\n",
    "geneFile = open(PANDAgeneListPath, \"w+\")\n",
    "\n",
    "for i in range(0, len(geneList3)):\n",
    "    newLine = geneList3[i] + \"\\t\" + \"\\n\"\n",
    "    geneFile.write(newLine)\n",
    "    \n",
    "geneFile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Terminal commands to run pypanda"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi.txt -o output_panda_all.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
