{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BreCaNet\n",
    "\n",
    "This Jupyter notebook goes through the steps to process the full data sets before running PANDA\n",
    "\n",
    "##### Gene Expression Data\n",
    "\n",
    "Came from the TCGA database: \n",
    "Settings for data: *Cases tab*\n",
    "    Primary site -> breast,\n",
    "    Disease type -> ductal and lobular neoplasms,\n",
    "    Gender -> female,\n",
    "    Race -> White\n",
    " \n",
    " *Files tab*\n",
    "    Experimental Strategy -> RNA-seq,\n",
    "    Workflow -> HTSeq-FPKM,\n",
    "    Project -> TCGA-BRCA\n",
    "    \n",
    "Click on the Download button and select Cart\n",
    "\n",
    "##### Clinical Data with IHC Receptor Status\n",
    "\n",
    "Using the same settings as for gathering the expression data, except instead of downloading the cart, download the manifest data. The manifest file is used with the supplemental R code and necessary to retrieve the clinical IHC information.\n",
    "\n",
    "Use the same settings in the TCGA repository on the Cases tab.\n",
    "\n",
    "*Files tab:*\n",
    "Data Category -> Clinical\n",
    "\n",
    "Data Format -> BCR Biotab\n",
    "\n",
    "Add to cart \n",
    "\n",
    "Download the nationawidechildrens.org_clinical_patient_brca.txt file\n",
    "\n",
    "##### TF Binding Motif Data\n",
    "\n",
    "GTex from Camila\n",
    "\n",
    "##### Protien-Protein Interaction (PPI) Data\n",
    "\n",
    "GTex from Camila\n",
    "\n",
    "##### PANDA algorithm\n",
    "\n",
    "https://netzoo.github.io"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gets PAM50 labels of TCGA samples\n",
    "\n",
    "###### How to get the labels for each sample:\n",
    "\n",
    "1) Use TCGA label of the folder of the expression file and match with the file_id in the metadata\n",
    "\n",
    "2) Use the case_id from the metadata that corresponds with the file_id to match the uuid in the clinical file\n",
    "\n",
    "3) Use the uuid to get the corresponding barcode, also in the clinical file\n",
    "\n",
    "4) Match the barcode from the clinical file to match with the barcode in the subtype label file to, finally, get the subtype of the sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import shutil\n",
    "import csv\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene expression paths\n",
    "dataPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/gdc_download_FPKM_832' #zipped file path\n",
    "dataNewFolder = '/Users/ursulawidocki/Desktop/BreCaNet/Data/processedTCGA_832' # folder for unzipped\n",
    "#PANDAexpPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_832.txt' # final exp file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unzips the expression data dn creates a folder (processedTCGA_832) of the unzipped files \n",
    "# Gets the codes of the patient folders and expression files\n",
    "\n",
    "folderIDs = list() # codes of the patient IDs on the folder\n",
    "indivList = list() # codes of the patients' expression file\n",
    "\n",
    "for fileName in os.listdir(dataPath):\n",
    "    if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "        \n",
    "    else:\n",
    "        folderIDs.append(fileName)\n",
    "        \n",
    "    newPath = dataPath + '/' + fileName\n",
    "    \n",
    "    # goes into the individual folder\n",
    "    for indivName in os.listdir(newPath):\n",
    "        if (indivName == '.DS_Store') or (indivName == 'MANIFEST.txt') or (indivName == 'annotations.txt'):\n",
    "            continue\n",
    "        filePath = newPath + '/' + indivName\n",
    "        \n",
    "        fileEnding = indivName[-3] + indivName[-2] + indivName[-1]\n",
    "        if(fileEnding == '.gz'):\n",
    "            with gzip.open(filePath, 'rb') as f_in:\n",
    "                newFilePath = dataNewFolder + '/' + indivName[0:-3]\n",
    "                indivList.append(indivName[0:-3])\n",
    "                \n",
    "                with open(newFilePath, 'wb') as f_out:\n",
    "                    shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes file of folder names\n",
    "\n",
    "folderIDPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/foldIDList_FPKM_832.txt'\n",
    "folderFile = open(folderIDPath, \"w+\")\n",
    "for i in range(0, 832):\n",
    "    newLineFolder = folderIDs[i] + \"\\n\"\n",
    "    \n",
    "    folderFile.write(newLineFolder)\n",
    "    \n",
    "folderFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Gets folder codes from expression data\n",
    "folderIDPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/foldIDList_FPKM_832.txt'\n",
    "folders = pd.read_csv(folderIDPath, delimiter = '\\t', header = None)\n",
    "folders_list = list(folders[0])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "with open(metaPath, 'r') as f:\n",
    "    meta_dict = json.load(f)\n",
    "    \n",
    "print(type(meta_dict[0]))\n",
    "\n",
    "def getList(dict): \n",
    "    return dict.keys() \n",
    "      \n",
    "# Driver program \n",
    "print(getList(meta_dict[0]))\n",
    "\n",
    "print(meta_dict[0].get('associated_entities'))\n",
    "\n",
    "#meta_dict\n",
    "\n",
    "#n = 0\n",
    "#with open(metaPath, \"r\") as file:\n",
    "    #for line in file.readlines():\n",
    "        #if (line[0:3] == \"   \") and (\"case_id\" in line):\n",
    "            #n+=1\n",
    "            \n",
    "#print(n)\n",
    "\n",
    "## new plan:\n",
    "\n",
    "for each dictionary in list:\n",
    "    get file id from key=file_id\n",
    "    get case_id from value of associated_entities\n",
    "    \n",
    " but first see how many you can get using if file in filders_list"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# if the line does not have a tab infront of \"file_id\", add to the list\n",
    "# if the line has "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "714\n",
      "714\n",
      "714\n"
     ]
    }
   ],
   "source": [
    "## Reads in metadata\n",
    "# gets file_id and case_id\n",
    "\n",
    "metaPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/metadata.cart.FPKM_832.json'\n",
    "\n",
    "file_id = list()\n",
    "case_id = list()\n",
    "with open(metaPath, \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        \n",
    "        if \"file_id\" in line:\n",
    "            item = line.strip().split(\"\\\"\")[3]\n",
    "            #print(item)\n",
    "                \n",
    "        if(\"case_id\" in line) and (line.strip().split(\"\\\"\")[3] not in case_id):\n",
    "            case = line.strip().split(\"\\\"\")[3]\n",
    "            case_id.append(case)\n",
    "            file_id.append(item)\n",
    "\n",
    "print(len(set(file_id)))\n",
    "print(len(case_id))\n",
    "print(len(set(case_id)))\n",
    "#print(case_id[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1097\n",
      "1097\n"
     ]
    }
   ],
   "source": [
    "## Gets uuid and barcodes of the patients from clinical data\n",
    "\n",
    "# reads in clinical information to link patients with their barcode\n",
    "clinicalPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/gdc_download_clinical_832/8162d394-8b64-4da2-9f5b-d164c54b9608/nationwidechildrens.org_clinical_patient_brca.txt'\n",
    "clinical_df = pd.read_csv(clinicalPath, delimiter = '\\t')\n",
    "\n",
    "# removes rows that are not necessary to match data\n",
    "clinical_df = clinical_df.drop(clinical_df.index[0])\n",
    "clinical_df = clinical_df.drop(clinical_df.index[0])\n",
    "clinical_df.head()\n",
    "codes_df = clinical_df[['bcr_patient_uuid', 'bcr_patient_barcode']]\n",
    "#codes_df.head()\n",
    "\n",
    "uuid_list = list(codes_df['bcr_patient_uuid'])\n",
    "barcode_list = list(codes_df['bcr_patient_barcode'])\n",
    "\n",
    "print(len(uuid_list))\n",
    "print(len(barcode_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1087\n",
      "1087\n"
     ]
    }
   ],
   "source": [
    "## Reads in PAM50 labels\n",
    "\n",
    "subtype_path = \"/Users/ursulawidocki/Desktop/BreCaNet/Code/TCGA_subtypes.txt\"\n",
    "subtypes = pd.read_csv(subtype_path, delimiter = \" \")\n",
    "subtypes.head()\n",
    "\n",
    "pt_barcodes_list = list(subtypes['brca_subtypes.patient'])\n",
    "subtype_list = list(subtypes['brca_subtypes.BRCA_Subtype_PAM50'])\n",
    "\n",
    "print(len(pt_barcodes_list))\n",
    "print(len(subtype_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Matching data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "700\n"
     ]
    }
   ],
   "source": [
    "## First, match the folder with the file_ids to get the corresponding cases from the metadata\n",
    "\n",
    "match_cases = list()\n",
    "\n",
    "for folder in folders_list:\n",
    "    for file in file_id:\n",
    "        if folder == file:\n",
    "            ind = file_id.index(folder)\n",
    "            match_cases.append(case_id[ind])\n",
    "            break\n",
    "\n",
    "print(len(match_cases)) # should be equivalent to number of samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "832\n"
     ]
    }
   ],
   "source": [
    "## Matches case_ids from metadata with uuid and patient barcodes from clinical data\n",
    "\n",
    "match_barcodes = list()\n",
    "\n",
    "for case in match_cases:\n",
    "    for ptID in uuid_list:\n",
    "        if case == ptID.lower():\n",
    "            ind = uuid_list.index(ptID)\n",
    "            match_barcodes.append(barcode_list[ind])\n",
    "            break\n",
    "            \n",
    "print(len(match_barcodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "832\n"
     ]
    }
   ],
   "source": [
    "## Matches subtypes to their barcodes from the TCGA data\n",
    "\n",
    "match_subtypes = list()\n",
    "\n",
    "for barcode in match_barcodes:\n",
    "    ind = pt_barcodes_list.index(barcode)\n",
    "    match_subtypes.append(subtype_list[ind])\n",
    "    \n",
    "print(len(match_subtypes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "{'LumA', 'Normal', 'LumB', 'Her2', 'Basal'}\n"
     ]
    }
   ],
   "source": [
    "print(len(set(match_subtypes))) # there is only this many subtypes among my samples\n",
    "print((set(match_subtypes)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now, sort the folder IDs into their subtypes\n",
    "\n",
    "Notice that the order of folders_list is in the same order as match_subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## uses index of subtype label to sort folder IDs into subtypes\n",
    "\n",
    "# Luminal A samples\n",
    "lumA_samples = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"LumA\":\n",
    "        lumA_samples.append(folders_list[i])\n",
    "\n",
    "# Luminal B samples\n",
    "lumB_samples = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"LumB\":\n",
    "        lumB_samples.append(folders_list[i])\n",
    "\n",
    "# Basal samples\n",
    "basal_samples = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"Basal\":\n",
    "        basal_samples.append(folders_list[i])\n",
    "\n",
    "# Her2 samples\n",
    "her2_samples = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"Her2\":\n",
    "        her2_samples.append(folders_list[i])\n",
    "\n",
    "# Normal samples\n",
    "norm_samples = list()\n",
    "for i in range(0,len(match_subtypes)):\n",
    "    if match_subtypes[i]==\"Normal\":\n",
    "        norm_samples.append(folders_list[i])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now that the expression data is sorted, normalize and filter before running PANDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "60483\n"
     ]
    }
   ],
   "source": [
    "# gets list of genes that are in the file\n",
    "\n",
    "geneList = list() # list of all genes that are noted in the expression files\n",
    "\n",
    "with open(dataNewFolder + '/' + indivList[0], \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        potentialGene = line.strip().split(\"\\t\")[0]\n",
    "        \n",
    "        if potentialGene[0:3] == \"ENS\":\n",
    "            newGeneName = potentialGene.strip().split(\".\")[0]\n",
    "            geneList.append(newGeneName)\n",
    "print(len(geneList))\n",
    "\n",
    "geneList_lumA = geneList\n",
    "geneList_lumB = geneList\n",
    "geneList_basal = geneList\n",
    "geneList_her2 = geneList\n",
    "geneList_norm = geneList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60483, 28)\n"
     ]
    }
   ],
   "source": [
    "## Reads through the unzipped expression files and creates the expression matrices depending on how the file\n",
    "# name matches with the folder IDs of each subtype\n",
    "\n",
    "lumA_exp = np.zeros((len(geneList), len(lumA_samples))) # exp matrix of LumA samples\n",
    "lumB_exp = np.zeros((len(geneList), len(lumB_samples))) # exp matrix of LumB samples\n",
    "basal_exp = np.zeros((len(geneList), len(basal_samples))) # exp matrix of basal samples\n",
    "her2_exp = np.zeros((len(geneList), len(her2_samples))) # exp matrix of her2 samples\n",
    "norm_exp = np.zeros((len(geneList), len(norm_samples))) # exp matrix of LumA samples\n",
    "\n",
    "print(norm_exp.shape) # just to make sure dimensions are correct\n",
    " \n",
    "lumA_j = 0 # index of the column (sample)index of the column (sample)\n",
    "lumB_j = 0\n",
    "basal_j = 0\n",
    "her2_j = 0\n",
    "norm_j = 0\n",
    "for fileName in os.listdir(dataNewFolder):\n",
    "    if fileName == '.DS_Store' or fileName == 'MANIFEST.txt':\n",
    "        continue\n",
    "    \n",
    "    if fileName in indivList:\n",
    "        \n",
    "        temp_folder = folderIDs[indivList.index(fileName)]\n",
    "        readFile = dataNewFolder + '/' + fileName\n",
    "            \n",
    "        # adds to luminal A expression matrix\n",
    "        if temp_folder in lumA_samples:\n",
    "            \n",
    "            lumA_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    lumA_exp[lumA_i, lumA_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    lumA_i += 1\n",
    "                    \n",
    "                lumA_j += 1\n",
    "                        \n",
    "        # adds to luminal B expression matrix\n",
    "        if temp_folder in lumB_samples:\n",
    "            \n",
    "            lumB_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    lumB_exp[lumB_i, lumB_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    lumB_i += 1\n",
    "                    \n",
    "                lumB_j += 1\n",
    "        \n",
    "        # adds to basal expression matrix\n",
    "        if temp_folder in basal_samples:\n",
    "            \n",
    "            basal_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    basal_exp[basal_i, basal_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    basal_i += 1\n",
    "                    \n",
    "                basal_j += 1\n",
    "        \n",
    "        # adds to her2 expression matrix\n",
    "        if temp_folder in her2_samples:\n",
    "            \n",
    "            her2_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    her2_exp[her2_i, her2_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    her2_i += 1\n",
    "                    \n",
    "                her2_j += 1\n",
    "        \n",
    "        # adds to norm expression matrix\n",
    "        if temp_folder in norm_samples:\n",
    "            \n",
    "            norm_i = 0\n",
    "            with open(readFile, \"r\") as file:\n",
    "                for line in file.readlines():\n",
    "                    norm_exp[norm_i, norm_j] = float(line.strip().split(\"\\t\")[1])\n",
    "                    norm_i += 1\n",
    "                    \n",
    "                norm_j += 1\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(49832, 28)\n",
      "OLD  60483\n",
      "ZEROS  10651\n",
      "NEW  49832\n",
      "49832\n"
     ]
    }
   ],
   "source": [
    "## Removes zero rows from expression data\n",
    "\n",
    "# Luminal A data\n",
    "rowList_lumA = list()\n",
    "for i in range(0, len(lumA_exp)):\n",
    "    if np.sum(lumA_exp[i,:]) == 0.0:\n",
    "        rowList_lumA.append(i)\n",
    "\n",
    "lumA_exp2 = np.zeros(((len(lumA_exp)-len(rowList_lumA)), len(lumA_samples))) # matrix without zero rows\n",
    "geneList_lumA2 = list()\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(lumA_exp)):\n",
    "    if j in rowList_lumA:\n",
    "        j+=1\n",
    "    else:\n",
    "        lumA_exp2[r,:] = lumA_exp[j,:]\n",
    "        geneList_lumA2.append(geneList_lumA[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "# Luminal B data\n",
    "rowList_lumB = list()\n",
    "for i in range(0, len(lumB_exp)):\n",
    "    if np.sum(lumB_exp[i,:]) == 0.0:\n",
    "        rowList_lumB.append(i)\n",
    "\n",
    "lumB_exp2 = np.zeros(((len(lumB_exp)-len(rowList_lumB)), len(lumB_samples))) # matrix without zero rows\n",
    "geneList_lumB2 = list()\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(lumB_exp)):\n",
    "    if j in rowList_lumB:\n",
    "        j+=1\n",
    "    else:\n",
    "        lumB_exp2[r,:] = lumB_exp[j,:]\n",
    "        geneList_lumB2.append(geneList_lumB[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "# Basal data\n",
    "rowList_basal = list()\n",
    "for i in range(0, len(basal_exp)):\n",
    "    if np.sum(basal_exp[i,:]) == 0.0:\n",
    "        rowList_basal.append(i)\n",
    "\n",
    "basal_exp2 = np.zeros(((len(basal_exp)-len(rowList_basal)), len(basal_samples))) # matrix without zero rows\n",
    "geneList_basal2 = list()\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(basal_exp)):\n",
    "    if j in rowList_basal:\n",
    "        j+=1\n",
    "    else:\n",
    "        basal_exp2[r,:] = basal_exp[j,:]\n",
    "        geneList_basal2.append(geneList_basal[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "# HER2 data\n",
    "rowList_her2 = list()\n",
    "for i in range(0, len(her2_exp)):\n",
    "    if np.sum(her2_exp[i,:]) == 0.0:\n",
    "        rowList_her2.append(i)\n",
    "\n",
    "her2_exp2 = np.zeros(((len(her2_exp)-len(rowList_her2)), len(her2_samples))) # matrix without zero rows\n",
    "geneList_her2_2 = list()\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(her2_exp)):\n",
    "    if j in rowList_her2:\n",
    "        j+=1\n",
    "    else:\n",
    "        her2_exp2[r,:] = her2_exp[j,:]\n",
    "        geneList_her2_2.append(geneList_her2[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "\n",
    "# Normal data\n",
    "rowList_norm = list()\n",
    "for i in range(0, len(norm_exp)):\n",
    "    if np.sum(norm_exp[i,:]) == 0.0:\n",
    "        rowList_norm.append(i)\n",
    "\n",
    "norm_exp2 = np.zeros(((len(norm_exp)-len(rowList_norm)), len(norm_samples))) # matrix without zero rows\n",
    "print(norm_exp2.shape)\n",
    "geneList_norm2 = list()\n",
    "\n",
    "print(\"OLD \", len(norm_exp))\n",
    "print(\"ZEROS \", len(rowList_norm))\n",
    "print(\"NEW \", len(norm_exp2))\n",
    "\n",
    "j = 0 # index in old matrix\n",
    "r = 0 # index in new matrix\n",
    "while (j < len(norm_exp)):\n",
    "    if j in rowList_norm:\n",
    "        j+=1\n",
    "    else:\n",
    "        norm_exp2[r,:] = norm_exp[j,:]\n",
    "        geneList_norm2.append(geneList_norm[j])\n",
    "        j+=1\n",
    "        r+=1\n",
    "        \n",
    "print(len(geneList_norm2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## log(TPM) normalization\n",
    "# TPMi = (FPKMi / sum(FPKMj)) * 10^6, where i is the gene and j is the sample\n",
    "# logTPM = log(TPM + 1)\n",
    "\n",
    "# Luminal A\n",
    "for j in range(0, len(lumA_samples)):\n",
    "    sumPat = np.sum(lumA_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(lumA_exp2)):\n",
    "        lumA_exp2[i,j] = np.log((lumA_exp2[i,j] / sumPat) * (10**6) + 1)\n",
    "        \n",
    "# Luminal B\n",
    "for j in range(0, len(lumB_samples)):\n",
    "    sumPat = np.sum(lumB_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(lumB_exp2)):\n",
    "        lumB_exp2[i,j] = np.log((lumB_exp2[i,j] / sumPat) * (10**6) + 1)\n",
    "        \n",
    "# Basal\n",
    "for j in range(0, len(basal_samples)):\n",
    "    sumPat = np.sum(basal_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(basal_exp2)):\n",
    "        basal_exp2[i,j] = np.log((basal_exp2[i,j] / sumPat) * (10**6) + 1)\n",
    "        \n",
    "# HER2\n",
    "for j in range(0, len(her2_samples)):\n",
    "    sumPat = np.sum(her2_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(her2_exp2)):\n",
    "        her2_exp2[i,j] = np.log((her2_exp2[i,j] / sumPat) * (10**6) + 1)\n",
    "        \n",
    "# Normal\n",
    "for j in range(0, len(norm_samples)):\n",
    "    sumPat = np.sum(norm_exp2, axis = 0)[j]\n",
    "    \n",
    "    for i in range(0, len(norm_exp2)):\n",
    "        norm_exp2[i,j] = np.log((norm_exp2[i,j] / sumPat) * (10**6) + 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Extracts TFBM information\n",
    "# keeps track of which TF and genes have relations as per TFBM file\n",
    "\n",
    "motifPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/motif.txt' # original motif file\n",
    "\n",
    "motifTF_lumA = list() # list of TFs interacting to genes that are expressed\n",
    "motifGenes_lumA = list() # list of genes in expression file with TFs\n",
    "motifTF_lumB = list()\n",
    "motifGenes_lumB = list()\n",
    "motifTF_basal = list()\n",
    "motifGenes_basal = list()\n",
    "motifTF_her2 = list()\n",
    "motifGenes_her2 = list()\n",
    "motifTF_norm = list()\n",
    "motifGenes_norm = list()\n",
    "\n",
    "with open(motifPath, \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        if line.strip().split(\"\\t\")[2] == \"1\":\n",
    "            \n",
    "            # gets motif data for Luminal A\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_lumA2:\n",
    "                motifTF_lumA.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_lumA.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "            # gets motif data for Luminal B\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_lumB2:\n",
    "                motifTF_lumB.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_lumB.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "            # gets motif data for Basal\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_basal:\n",
    "                motifTF_basal.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_basal.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "            # gets motif data for HER2\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_her2:\n",
    "                motifTF_her2.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_her2.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "            # gets motif data for Normal\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_norm:\n",
    "                motifTF_norm.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_norm.append(line.strip().split(\"\\t\")[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1792566\n",
      "1792566\n",
      "1790056\n",
      "1790056\n",
      "1806322\n",
      "1806322\n",
      "1806322\n",
      "1806322\n",
      "1806322\n",
      "1806322\n"
     ]
    }
   ],
   "source": [
    "print(len(motifTF_lumA))\n",
    "print(len(motifGenes_lumA))\n",
    "\n",
    "print(len(motifTF_lumB))\n",
    "print(len(motifGenes_lumB))\n",
    "\n",
    "print(len(motifTF_basal))\n",
    "print(len(motifGenes_basal))\n",
    "\n",
    "print(len(motifTF_her2))\n",
    "print(len(motifGenes_her2))\n",
    "\n",
    "print(len(motifTF_norm))\n",
    "print(len(motifGenes_norm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28936\n",
      "26929\n",
      "27078\n",
      "24096\n",
      "21579\n"
     ]
    }
   ],
   "source": [
    "## Checks if there are genes expressed and not in the motif file\n",
    "# if not, then doesn't add to updated list of genes in the exp matrix\n",
    "# and makes note of it to remove later\n",
    "\n",
    "# Luminal A\n",
    "removeGene_lumA = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_lumA2)):\n",
    "    if geneList_lumA2[i] not in motifGenes_lumA:\n",
    "        removeGene_lumA.append(i)\n",
    "        \n",
    "print(len(removeGene_lumA))\n",
    "\n",
    "## Luminal B\n",
    "removeGene_lumB = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_lumB2)):\n",
    "    if geneList_lumB2[i] not in motifGenes_lumB:\n",
    "        removeGene_lumB.append(i)\n",
    "        \n",
    "print(len(removeGene_lumB))\n",
    "\n",
    "# Basal\n",
    "removeGene_basal = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_basal2)):\n",
    "    if geneList_basal2[i] not in motifGenes_basal:\n",
    "        removeGene_basal.append(i)\n",
    "        \n",
    "print(len(removeGene_basal))\n",
    "\n",
    "# HER2\n",
    "removeGene_her2 = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_her2_2)):\n",
    "    if geneList_her2_2[i] not in motifGenes_her2:\n",
    "        removeGene_her2.append(i)\n",
    "        \n",
    "print(len(removeGene_her2))\n",
    "\n",
    "# Normal\n",
    "removeGene_norm = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_norm2)):\n",
    "    if geneList_norm2[i] not in motifGenes_norm:\n",
    "        removeGene_norm.append(i)\n",
    "        \n",
    "print(len(removeGene_norm))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28253, 28)\n",
      "28253\n"
     ]
    }
   ],
   "source": [
    "## Removes the rows and genes that are not in the motif file\n",
    "\n",
    "# Luminal A\n",
    "lumA_exp3 = np.zeros((len(lumA_exp2)-len(removeGene_lumA), len(lumA_samples)))\n",
    "geneList_lumA3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(lumA_exp2)):\n",
    "    if i not in removeGene_lumA:\n",
    "        lumA_exp3[r,:] = lumA_exp2[i,:]\n",
    "        geneList_lumA3.append(geneList_lumA2[i])\n",
    "        r+=1\n",
    "\n",
    "# Luminal B\n",
    "lumB_exp3 = np.zeros((len(lumB_exp2)-len(removeGene_lumB), len(lumB_samples)))\n",
    "geneList_lumB3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(lumB_exp2)):\n",
    "    if i not in removeGene_lumB:\n",
    "        lumB_exp3[r,:] = lumB_exp2[i,:]\n",
    "        geneList_lumB3.append(geneList_lumB2[i])\n",
    "        r+=1\n",
    "\n",
    "# Basal\n",
    "basal_exp3 = np.zeros((len(basal_exp2)-len(removeGene_basal), len(basal_samples)))\n",
    "geneList_basal3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(basal_exp2)):\n",
    "    if i not in removeGene_basal:\n",
    "        basal_exp3[r,:] = basal_exp2[i,:]\n",
    "        geneList_basal3.append(geneList_basal2[i])\n",
    "        r+=1\n",
    "        \n",
    "# HER2\n",
    "her2_exp3 = np.zeros((len(her2_exp2)-len(removeGene_her2), len(her2_samples)))\n",
    "geneList_her2_3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(her2_exp2)):\n",
    "    if i not in removeGene_her2:\n",
    "        her2_exp3[r,:] = her2_exp2[i,:]\n",
    "        geneList_her2_3.append(geneList_her2_2[i])\n",
    "        r+=1\n",
    "        \n",
    "# Normal\n",
    "norm_exp3 = np.zeros((len(norm_exp2)-len(removeGene_norm), len(norm_samples)))\n",
    "geneList_norm3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(norm_exp2)):\n",
    "    if i not in removeGene_norm:\n",
    "        norm_exp3[r,:] = norm_exp2[i,:]\n",
    "        geneList_norm3.append(geneList_norm2[i])\n",
    "        r+=1\n",
    "        \n",
    "print(norm_exp3.shape)\n",
    "print(len(geneList_norm3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "29374\n",
      "28552\n",
      "822\n",
      "True\n",
      "29374\n",
      "28253\n",
      "1121\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "## Removes genes the are not expressed but in motif file\n",
    "\n",
    "# gets all indeces of an element(gene) in a list(motifGenes)\n",
    "def getIndeces(listOfElements, element):\n",
    "    indexPosList = []\n",
    "    indexPos = 0\n",
    "    while True:\n",
    "        try:\n",
    "            # Search for item in list from indexPos to the end of list\n",
    "            indexPos = listOfElements.index(element, indexPos)\n",
    "            # Add the index position in list\n",
    "            indexPosList.append(indexPos)\n",
    "            indexPos += 1\n",
    "        except ValueError as e:\n",
    "            break\n",
    " \n",
    "    return indexPosList\n",
    "\n",
    "# Luminal A\n",
    "if set(motifGenes_lumA) != set(geneList_lumA3):\n",
    "    print(len(set(motifGenes_lumA)))\n",
    "    print(len(set(geneList_lumA3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_lumA) - set(geneList_lumA3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_lumA, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_lumA.remove(motifGenes_lumA[geneInds[i] - i])\n",
    "            motifTF_lumA.remove(motifTF_lumA[geneInds[i] - i])\n",
    "\n",
    "# Luminal B\n",
    "if set(motifGenes_lumB) != set(geneList_lumB3):\n",
    "    print(len(set(motifGenes_lumB)))\n",
    "    print(len(set(geneList_lumB3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_lumB) - set(geneList_lumB3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_lumB, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_lumB.remove(motifGenes_lumB[geneInds[i] - i])\n",
    "            motifTF_lumB.remove(motifTF_lumB[geneInds[i] - i])\n",
    "\n",
    "# Basal\n",
    "if set(motifGenes_basal) != set(geneList_basal3):\n",
    "    print(len(set(motifGenes_basal)))\n",
    "    print(len(set(geneList_basal3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_basal) - set(geneList_basal3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_basal, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_basal.remove(motifGenes_basal[geneInds[i] - i])\n",
    "            motifTF_basal.remove(motifTF_basal[geneInds[i] - i])\n",
    "            \n",
    "print(set(motifGenes_basal)== set(geneList_basal3))\n",
    "\n",
    "# HER2\n",
    "if set(motifGenes_her2) != set(geneList_her2_3):\n",
    "    print(len(set(motifGenes_her2)))\n",
    "    print(len(set(geneList_her2_3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_her2) - set(geneList_her2_3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_her2, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_her2.remove(motifGenes_her2[geneInds[i] - i])\n",
    "            motifTF_her2.remove(motifTF_her2[geneInds[i] - i])\n",
    "            \n",
    "print(set(motifGenes_her2)== set(geneList_her2_3))\n",
    "\n",
    "# Normal\n",
    "if set(motifGenes_norm) != set(geneList_norm3):\n",
    "    print(len(set(motifGenes_norm)))\n",
    "    print(len(set(geneList_norm3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_norm) - set(geneList_norm3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_norm, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_norm.remove(motifGenes_norm[geneInds[i] - i])\n",
    "            motifTF_norm.remove(motifTF_norm[geneInds[i] - i])\n",
    "            \n",
    "print(set(motifGenes_norm)== set(geneList_norm3))\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## DEBUGGING\n",
    "# Basal pipeline\n",
    "\n",
    "# Calc first motif objects\n",
    "motifTF_basal = list()\n",
    "motifGenes_basal = list()\n",
    "\n",
    "with open(motifPath, \"r\") as file:\n",
    "    for line in file.readlines():\n",
    "        if line.strip().split(\"\\t\")[2] == \"1\":\n",
    "            # gets motif data for Basal\n",
    "            if line.strip().split(\"\\t\")[1] in geneList_basal:\n",
    "                motifTF_basal.append(line.strip().split(\"\\t\")[0])\n",
    "                motifGenes_basal.append(line.strip().split(\"\\t\")[1])\n",
    "                \n",
    "print(len(motifTF_basal))\n",
    "print(len(motifGenes_basal))\n",
    "\n",
    "# removes from expression if not in motif list\n",
    "removeGene_basal = list() # list of indeces to remove from expressed genes\n",
    "for i in range(0, len(geneList_basal2)):\n",
    "    if geneList_basal2[i] not in motifGenes_basal:\n",
    "        removeGene_basal.append(i)\n",
    "        \n",
    "print(len(removeGene_basal))\n",
    "\n",
    "basal_exp3 = np.zeros((len(basal_exp2)-len(removeGene_basal), len(basal_samples)))\n",
    "geneList_basal3 = list() # list of genes that are in motif file and the motifs are expressed genes\n",
    "\n",
    "r = 0 # index of row in allTesting3\n",
    "for i in range(0, len(basal_exp2)):\n",
    "    if i not in removeGene_basal:\n",
    "        basal_exp3[r,:] = basal_exp2[i,:]\n",
    "        geneList_basal3.append(geneList_basal2[i])\n",
    "        r+=1\n",
    "        \n",
    "        \n",
    "# logTPM        \n",
    "for j in range(0, len(basal_samples)):\n",
    "    for i in range(0, len(geneList_basal3)):\n",
    "        basal_exp3[i,j] = np.log(basal_exp3[i,j] + 1)\n",
    "        \n",
    "# removes motifGenes if not expressed\n",
    "if set(motifGenes_basal) != set(geneList_basal3):\n",
    "    print(len(set(motifGenes_basal)))\n",
    "    print(len(set(geneList_basal3)))\n",
    "    \n",
    "    diff = list(set(motifGenes_basal) - set(geneList_basal3))\n",
    "    print(len(diff))\n",
    "    for gene in diff:\n",
    "        geneInds = getIndeces(motifGenes_basal, gene)\n",
    "\n",
    "        for i in range(0, len(geneInds)):\n",
    "            motifGenes_basal.remove(motifGenes_basal[geneInds[i] - i])\n",
    "            motifTF_basal.remove(motifTF_basal[geneInds[i] - i])\n",
    "            \n",
    "print(set(motifGenes_basal)== set(geneList_basal3))\n",
    "\n",
    "# These should be the same\n",
    "print(len(motifTF_basal))\n",
    "print(len(motifGenes_basal))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1790318\n",
      "1790318\n",
      "1782492\n",
      "1782492\n",
      "1772784\n",
      "1772784\n"
     ]
    }
   ],
   "source": [
    "print(len(motifTF_basal))\n",
    "print(len(motifGenes_basal))\n",
    "\n",
    "print(len(motifTF_her2))\n",
    "print(len(motifGenes_her2))\n",
    "\n",
    "print(len(motifTF_norm))\n",
    "print(len(motifGenes_norm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Reads in the ppi data and extracts the proteins \n",
    "# About 4 hrs to run\n",
    "\n",
    "ppiPath = '/Users/ursulawidocki/Desktop/BreCaNet/Data/ppi2015_freeze.txt'\n",
    "\n",
    "protein1_lumA = list()\n",
    "protein2_lumA = list()\n",
    "protein1_lumB = list()\n",
    "protein2_lumB = list()\n",
    "protein1_basal = list()\n",
    "protein2_basal = list()\n",
    "protein1_her2 = list()\n",
    "protein2_her2 = list()\n",
    "protein1_norm = list()\n",
    "protein2_norm = list()\n",
    "\n",
    "with open(ppiPath, \"r\") as file:\n",
    "    next(file) #skips the first line since that is just a header\n",
    "    for line in file.readlines():\n",
    "        if line.strip().split(\"\\t\")[0] in motifTF_lumA:\n",
    "            protein1_lumA.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_lumA.append(line.strip().split(\"\\t\")[1])\n",
    "        \n",
    "        if line.strip().split(\"\\t\")[0] in motifTF_lumB:\n",
    "            protein1_lumB.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_lumB.append(line.strip().split(\"\\t\")[1])\n",
    "            \n",
    "        if line.strip().split(\"\\t\")[0] in motifTF_basal:\n",
    "            protein1_basal.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_basal.append(line.strip().split(\"\\t\")[1])\n",
    "            \n",
    "        if line.strip().split(\"\\t\")[0] in motifTF_her2:\n",
    "            protein1_her2.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_her2.append(line.strip().split(\"\\t\")[1])\n",
    "            \n",
    "        if line.strip().split(\"\\t\")[0] in motifTF_norm:\n",
    "            protein1_norm.append(line.strip().split(\"\\t\")[0])\n",
    "            protein2_norm.append(line.strip().split(\"\\t\")[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Same!!\n",
      "Same!!\n",
      "Same!!\n",
      "Same!!\n",
      "Same!!\n"
     ]
    }
   ],
   "source": [
    "# Check to see if the two protein lists have the same proteins or not\n",
    "\n",
    "if(set(protein1_lumA) == set(protein2_lumA)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "if(set(protein1_lumB) == set(protein2_lumB)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "if(set(protein1_basal) == set(protein2_basal)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "if(set(protein1_her2) == set(protein2_her2)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "if(set(protein1_norm) == set(protein2_norm)):\n",
    "    print(\"Same!!\")\n",
    "    \n",
    "# all proteins sets contain the same proteins"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## DO NOT NEED if all varaibles with gene names are ENS IDs\n",
    "\n",
    "## Converts ENS IDs in geneLists to Entrez\n",
    "\n",
    "convertFilePath = \"/Users/ursulawidocki/Desktop/BreCaNet/Data/geneNames.txt\"\n",
    "\n",
    "# gets Entrez IDs of genes \n",
    "with open(convertFilePath, \"r\") as file:\n",
    "    next(file)\n",
    "    for line in file.readlines():\n",
    "        #print(line.strip().split(\"\\t\")[0], line.strip().split(\"\\t\")[1])\n",
    "        ensID = line.strip().split(\"\\t\")[0]\n",
    "        \n",
    "        if ensID in geneList_lumA3:\n",
    "            geneList_lumA3[geneList_lumA3.index(ensID)] = line.strip().split(\"\\t\")[1]\n",
    "            \n",
    "        if ensID in geneList_lumB3:\n",
    "            geneList_lumB3[geneList_lumB3.index(ensID)] = line.strip().split(\"\\t\")[1]\n",
    "            \n",
    "        if ensID in geneList_basal3:\n",
    "            geneList_basal3[geneList_basal3.index(ensID)] = line.strip().split(\"\\t\")[1]\n",
    "            \n",
    "        if ensID in geneList_her2_3:\n",
    "            geneList_her2_3[geneList_her2_3.index(ensID)] = line.strip().split(\"\\t\")[1]\n",
    "            \n",
    "        if ensID in geneList_norm3:\n",
    "            geneList_norm3[geneList_norm3.index(ensID)] = line.strip().split(\"\\t\")[1]\n",
    "\n",
    "# locates duplicate genes\n",
    "print(len(geneList_lumA3) - len(set(geneList_lumA3))) \n",
    "print(len(geneList_lumB3) - len(set(geneList_lumB3)))\n",
    "print(len(geneList_basal3) - len(set(geneList_basal3)))\n",
    "print(len(geneList_her2_3) - len(set(geneList_her2_3)))\n",
    "print(len(geneList_norm3) - len(set(geneList_norm3)))\n",
    "\n",
    "# there are no duplicates present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes the expression filess\n",
    "\n",
    "# Luminal A\n",
    "PANDAexpPath_lumA = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_lumA.txt'\n",
    "geneExpFile = open(PANDAexpPath_lumA,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_lumA3)):\n",
    "    newLine = geneList_lumA3[i]\n",
    "    \n",
    "    for j in range(0, len(lumA_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(lumA_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()\n",
    "\n",
    "# Luminal B\n",
    "PANDAexpPath_lumB = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_lumB.txt'\n",
    "geneExpFile = open(PANDAexpPath_lumB,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_lumB3)):\n",
    "    newLine = geneList_lumB3[i]\n",
    "    \n",
    "    for j in range(0, len(lumB_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(lumB_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()\n",
    "\n",
    "# Basal\n",
    "PANDAexpPath_basal = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_basal.txt'\n",
    "geneExpFile = open(PANDAexpPath_basal,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_basal3)):\n",
    "    newLine = geneList_basal3[i]\n",
    "    \n",
    "    for j in range(0, len(basal_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(basal_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()\n",
    "\n",
    "# HER2\n",
    "PANDAexpPath_her2 = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_her2.txt'\n",
    "geneExpFile = open(PANDAexpPath_her2,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_her2_3)):\n",
    "    newLine = geneList_her2_3[i]\n",
    "    \n",
    "    for j in range(0, len(her2_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(her2_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()\n",
    "\n",
    "# Normal\n",
    "PANDAexpPath_norm = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_norm.txt'\n",
    "geneExpFile = open(PANDAexpPath_norm,\"w+\")\n",
    "\n",
    "for i in range(0, len(geneList_norm3)):\n",
    "    newLine = geneList_norm3[i]\n",
    "    \n",
    "    for j in range(0, len(norm_samples)):\n",
    "        newLine = newLine + \"\\t\" + str(norm_exp3[i,j])\n",
    "        \n",
    "    newLine = newLine + \"\\n\"\n",
    "    geneExpFile.write(newLine)\n",
    "    \n",
    "geneExpFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes the motif files\n",
    "\n",
    "# Luminal A\n",
    "PANDAmotifPath_lumA = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_lumA.txt'\n",
    "motifFile = open(PANDAmotifPath_lumA, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_lumA)):\n",
    "    newLine = motifTF_lumA[i] + \"\\t\" + motifGenes_lumA[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()\n",
    "\n",
    "# Luminal B\n",
    "PANDAmotifPath_lumB = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_lumB.txt'\n",
    "motifFile = open(PANDAmotifPath_lumB, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_lumB)):\n",
    "    newLine = motifTF_lumB[i] + \"\\t\" + motifGenes_lumB[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()\n",
    "\n",
    "# Basal\n",
    "PANDAmotifPath_basal = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_basal.txt'\n",
    "motifFile = open(PANDAmotifPath_basal, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_basal)):\n",
    "    newLine = motifTF_basal[i] + \"\\t\" + motifGenes_basal[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()\n",
    "\n",
    "# HER2\n",
    "PANDAmotifPath_her2 = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_her2.txt'\n",
    "motifFile = open(PANDAmotifPath_her2, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_her2)):\n",
    "    newLine = motifTF_her2[i] + \"\\t\" + motifGenes_her2[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()\n",
    "\n",
    "# Normal\n",
    "PANDAmotifPath_norm = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_norm.txt'\n",
    "motifFile = open(PANDAmotifPath_norm, \"w+\")\n",
    "\n",
    "for i in range(0, len(motifTF_norm)):\n",
    "    newLine = motifTF_norm[i] + \"\\t\" + motifGenes_norm[i] + \"\\t\" + \"1.000000\" + \"\\n\"\n",
    "    motifFile.write(newLine)\n",
    "\n",
    "motifFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Makes the ppi files\n",
    "\n",
    "# Luminal A\n",
    "PANDAppiPath_lumA = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_lumA.txt'\n",
    "ppiFile = open(PANDAppiPath_lumA, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_lumA)):\n",
    "    newLine = protein1_lumA[i] + \"\\t\" + protein2_lumA[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()\n",
    "\n",
    "# Luminal B\n",
    "PANDAppiPath_lumB = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_lumB.txt'\n",
    "ppiFile = open(PANDAppiPath_lumB, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_lumB)):\n",
    "    newLine = protein1_lumB[i] + \"\\t\" + protein2_lumB[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()\n",
    "\n",
    "# Basal\n",
    "PANDAppiPath_basal = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_basal.txt'\n",
    "ppiFile = open(PANDAppiPath_basal, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_basal)):\n",
    "    newLine = protein1_basal[i] + \"\\t\" + protein2_basal[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()\n",
    "\n",
    "# HER2\n",
    "PANDAppiPath_her2 = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_her2.txt'\n",
    "ppiFile = open(PANDAppiPath_her2, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_her2)):\n",
    "    newLine = protein1_her2[i] + \"\\t\" + protein2_her2[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()\n",
    "\n",
    "# Normal\n",
    "PANDAppiPath_norm = '/Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_norm.txt'\n",
    "ppiFile = open(PANDAppiPath_norm, \"w+\")\n",
    "\n",
    "for i in range(0, len(protein1_norm)):\n",
    "    newLine = protein1_norm[i] + \"\\t\" + protein2_norm[i] + \"\\t\" + \"1\" + \"\\n\"\n",
    "    ppiFile.write(newLine)\n",
    "\n",
    "ppiFile.close()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_lumA.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_lumA.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_lumA.txt -o output_panda_lumA.txt\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_lumB.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_lumB.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_lumB.txt -o output_panda_lumB.txt\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_basal.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_basal.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_basal.txt -o output_panda_basal.txt\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_her2.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_her2.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_her2.txt -o output_panda_her2.txt\n",
    "\n",
    "python run_panda.py -e /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAgeneExp_norm.txt -m /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAmotifs_norm.txt -p /Users/ursulawidocki/Desktop/BreCaNet/Data/PANDAinput/PANDAppi_norm.txt -o output_panda_norm.txt\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
